<?xml version="1.0"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<title>Claus Wilke, Christoph Adami 'The biology of digital organisms' (VX heaven)</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8"/>
<meta name="Author" content="Claus Wilke, Christoph Adami"/>
<meta name="KeyWords" lang="en" content="computer virus, virus, virii,vx, компьютерные вирусы, вирус, вири, Wilke, Claus; Adami, Christoph,biology of digital organisms, mutations, mutation, long, epistasis, life, programs, copy, populations, reaction, strain, instruction, computational, species, quasi, interactions"/>
<meta name="Description" content="Digital organisms are self-replicating computer programs that mutate and evolve. They can be thought of as a domesticated form of computer virus that lives in, and adapts to,a controlled environment.Digital organisms provide a unique opportunity with which to study evolutionary biology in a form of life that shares no ancestry with carbon-based life forms,and hence to distinguish general principles of evolution from historical accidents that are particular to biochemical life.In terms of the complexity of their evolutionary dynamics,digital organisms can be compared with biochemical viruses and bacteria.Recent studies of digital organisms have addressed long-term evolutionary adaptation and the growth of complexity in evolving systems,patterns of epistatic interactions in various genetic backgrounds,and quasi-species dynamics."/>
<script type="text/javascript">
//<![CDATA[
try{if (!window.CloudFlare) {var CloudFlare=[{verbose:0,p:0,byc:0,owlid:"cf",bag2:1,mirage2:0,oracle:0,paths:{cloudflare:"/cdn-cgi/nexp/dok3v=1613a3a185/"},atok:"047a5bcbf67431883fc9ed25fba33612",petok:"d0c9c6071474b264c3fddd9ea2ed8fb0258d7a04-1498755661-1800",zone:"vxheaven.org",rocket:"a",apps:{}}];document.write('<script type="text/javascript" src="//ajax.cloudflare.com/cdn-cgi/nexp/dok3v=85b614c0f6/cloudflare.min.js"><'+'\/script>');}}catch(e){};
//]]>
</script>
<link rel="icon" href="/favicon.ico" type="image/x-icon"/>
<link rel="shortcut icon" href="/favicon.ico" type="image/x-icon"/>
<link rel="stylesheet" type="text/css" href="/style.css"/><link rel="canonical" href="http://vxheaven.org/lib/mcw00.html"/>
<script type="text/rocketscript" data-rocketsrc="https://apis.google.com/js/plusone.js">{"parsetags": "explicit"}</script>
</head>
<body bgcolor="#dbc8a0" text="#302000" link="#225599" vlink="#113366">
<div class="s1">
<div style="float:right;"><a href="/lib/index.php?tbs=1"><img src="/img/max.gif" alt="Maximize"/></a></div> <form id="lf" style="margin: 0; float: right;" method="get" action="/index.php"><input type="hidden" name="action" value="set"/><select name="lang" onchange="javascript:document.getElementById('lf').submit();"><option value="ru">Русский</option><option selected="selected" value="en">English</option><option value="ua">Українська</option><option value="de">Deutsch</option><option value="es">Español</option><option value="fr">Fran&ccedil;ais</option><option value="it">Italiano</option><option value="pl">Polski</option></select></form>
<div style="float: right;"><div id="plusone"></div></div>
<script type="text/rocketscript">gapi.plusone.render("plusone", {"size":"small","count":"true"});</script>
<div style="float: right;" class="addthis_toolbox addthis_default_style">
<script type="text/rocketscript">var addthis_config = { ui_click: true }</script>
<a style="text-decoration: none; font-size: 10pt;" href="/?action=addthis" class="addthis_button_compact">Bookmark</a>
<script type="text/rocketscript" data-rocketsrc="http://s7.addthis.com/js/250/addthis_widget.js#username=herm1t"></script>
</div>
<div style="float: right;">
<script type="text/rocketscript" data-rocketsrc="http://www.google.com/cse/brand?form=cse-search-box&amp;lang=en"></script>
<form action="/search.php" id="cse-search-box">
<input type="hidden" name="cx" value="002577580816726040001:z9_irkorydo"/>
<input type="hidden" name="cof" value="FORID:10"/>
<input type="hidden" name="ie" value="UTF-8"/>
<input type="text" name="q" size="12" value=" "/>
<input type="submit" name="sa" value="Search"/>
</form>
</div><h1><a href="/" style="text-decoration: none; color: #000000;">VX Heaven</a></h1>
<span class="nav"><a href="/lib/">Library</a> <a href="/vl.php">Collection</a> <a href="/src.php">Sources</a> <a href="/vx.php?id=eidx">Engines</a> <a href="/vx.php?id=tidx">Constructors</a> <a href="/vx.php?id=sidx">Simulators</a> <a href="/vx.php?id=uidx">Utilities</a> <a href="/links.php">Links</a> <a href="/donate.php" style="color: #706020" id="donate">Donate</a> <a href="/forum" style="text-decoration: underline;">Forum</a> </span><br clear="all"/>
</div>
<div class="s2"><h1>The biology of digital organisms</h1><p><a href="/lib/?lang=en&amp;author=Wilke%2C%20Claus">Claus Wilke</a>, <a href="/lib/?lang=en&amp;author=Adami%2C%20Christoph">Christoph Adami</a><br/> <em>Trends in Ecology & Evolution, Vol.17, No.11, pp. 528-532</em><br/> <em>ISSN 0169-5347</em><br/> <em>November 2002</em></p><script type="text/rocketscript">var disqus_url = 'http://vxheaven.org/lib/mcw00.html';</script><img src="/img/pdf.gif" alt="PDF"/><a href="/lib/pdf/The%20biology%20of%20digital%20organisms.pdf">Download</a> PDF (41.94Kb) (You need to be registered on <a href="/forum">forum</a>)<br/>[<a style="" href="/lib/?lang=EN&amp;index=AI#mcw00">Back to index</a>] [<a href="/lib/mcw00.html#disqus_thread">Comments</a>]<br/> <form method="post" action=""><input type="hidden" name="pages" value="1"/><input type="submit" value="Turn on page numbers"/></form>
<address>
<p>Claus O. Wilke<br/>
Digital Life Laboratory, MC 136-93,<br/>
California Institute of Technology,<br/>
Pasadena, CA 91125, USA.<br/>
*e-mail: <a class="__cf_email__" href="/cdn-cgi/l/email-protection" data-cfemail="44332d282f21042725283021272c6a212031">[email&#160;protected]</a><script data-cfhash='f9e31' type="text/rocketscript">/* <![CDATA[ */!function(t,e,r,n,c,a,p){try{t=document.currentScript||function(){for(t=document.getElementsByTagName('script'),e=t.length;e--;)if(t[e].getAttribute('data-cfhash'))return t[e]}();if(t&&(c=t.previousSibling)){p=t.parentNode;if(a=c.getAttribute('data-cfemail')){for(e='',r='0x'+a.substr(0,2)|0,n=2;a.length-n;n+=2)e+='%'+('0'+('0x'+a.substr(n,2)^r).toString(16)).slice(-2);p.replaceChild(document.createTextNode(decodeURIComponent(e)),c)}p.removeChild(t)}}catch(u){}}()/* ]]> */</script></p>
<p>Christoph Adami<br/>
Jet Propulsion Laboratory, MC 126-347,<br/>
California Institute of Technology,<br/>
Pasadena, CA 91125, USA.</p>
</address>
<ul>
<li><a href="#c0">Long-term adaptation</a></li>
<li><a href="#c1">Epistatic interactions</a></li>
<li><a href="#c2">Quasi-species dynamics</a></li>
<li><a href="#c3">Future research directions</a>
<ul>
<li><a href="#c31">Digital life genetics</a></li>
<li><a href="#c32">Evolutionary ecology</a></li>
</ul></li>
<li><a href="#c4">Acknowledgments</a></li>
<li><a href="#c5">Conclusions</a></li>
<li><a href="#c6">Glossary</a></li>
<li><a href="#c7">References</a></li>
</ul>
<p><strong>Digital organisms are self-replicating computer programs that mutate and evolve. They can be thought of as a domesticated form of computer virus that lives in, and adapts to,a controlled environment.Digital organisms provide a unique opportunity with which to study evolutionary biology in a form of life that shares no ancestry with carbon-based life forms,and hence to distinguish general principles of evolution from historical accidents that are particular to biochemical life.In terms of the complexity of their evolutionary dynamics,digital organisms can be compared with biochemical viruses and bacteria.Recent studies of digital organisms have addressed long-term evolutionary adaptation and the growth of complexity in evolving systems,patterns of epistatic interactions in various genetic backgrounds,and quasi-species dynamics.</strong></p>
<p><em>Published online: 03 September 2002</em></p>
<p>Historically, evolutionary biology has generally been an observational and theoretical science. The experimental verification of evolutionary mechanisms is a challenging undertaking for several reasons: most organisms have comparatively long generation times; there are difficulties in determining important parameters, such as mutation rates or fitness values; and the large variances inherent in evolution lead to poor statistical significance in averages. Here, we highlight recent work on digital organisms, for which generation times are of the order of seconds and measurements can be taken with unprecedented accuracy. Digital organisms are self-replicating computer programs that live in a controlled environment. Unlike other computational approaches to studying evolution (such as genetic algorithms [1]or numerical simulations) digital organisms must explicitly create a copy of their own GENOME (see Glossary) to reproduce, and no particular genomic sequence is designated as the target or optimal sequence. Selection occurs because the environment in which the digital organisms live is space limited, that is, with the birth of a new organism an older one (typically chosen randomly) is removed from the population. Therefore, those organisms that produce more offspring replace less efficient replicators over time.</p>
<p>Mutations occur in digital organisms via explicit genomic errors (such as point mutations incurred during the copy process), or as <em>IMPLICIT MUTATIONS</em> that are the result of flawed copy algorithms. For example, an organism might skip part of its genome during replication, or replicate part of its genome twice. Copy mutations occur because the copying of a single <em>INSTRUCTION</em> in the genome has a certain probability of failure that results in a random instruction written into the daughter genome. Other <em>EXPLICIT MUTATIONS</em> are random changes in the genome of the organism that occur independently of the copy process (cosmic ray mutations), or random insertions and/or deletions of single instructions. The rates of explicit mutations are under the control of the researcher, whereas implicit mutations cannot typically be controlled.</p>
<p>Digital organisms have been studied for the past 12 years (Box 1), but only recently have evolutionary experiments with digital organisms reached a level of sophistication that is comparable to that of experiments with bacteria or viruses. What kind of question can be addressed with digital organisms? Clearly, because digital organisms live in a completely artificial world, every conclusion from a digital life experiment is potentially an artifact of the particular choices of that digital world (Box 2). However, this apparent weakness of digital biology is, at the same time, its biggest strength. By comparing results across wide ranges of parameter settings in the digital world, as well as with results from biochemical organisms and from mathematical theories, it is possible to disentangle general principles from effects that are peculiar to a particular model organism. Here, we highlight three distinct topics that have recently been addressed using digital organisms: the dynamics of long-term adaptation, the distribution of epistatic interactions among mutations, and quasi-species dynamics. Moreover, we discuss two areas of future research that appear to be promising, digital life genetics and evolutionary ecology.</p>
<h2><a name="c0"></a>Long-term adaptation</h2>
<p>One of the cornerstones of evolutionary biology is the influence of mutation and selection on organisms over long periods (of the order of thousands of generations or more), because darwinian theory predicts macroevolution and the emergence of novelty on that timescale. However, this is also one of the most difficult aspects to study, because of the long generation time of most model organisms. Macroevolutionary changes in biochemical organisms can only be studied through the history of domesticated animals and plants, or through the study of fossil data and molecular sequence similarities among species. However, these are purely observational approaches and do not allow manipulation of key parameters of the evolutionary process.Alternatively, one can study the long-term evolution of organisms with extremely short generation times, such as bacteria and viruses, an approach taken to new lengths by Lenski and co-workers [2–4].Yet, even for these microorganisms, it takes years to propagate populations through many thousands of generations.</p>
<p>For digital organisms, such a propagation can be achieved in a matter of days, and it is therefore not
 
surprising that experiments with digital organisms have traditionally focused on long-term evolutionary trends [5–12]. Early observations made in such experiments showed that evolution unfolds in an intermittent fashion [7,8]. Brief periods of rapid increases in fitness are interspersed with long periods during which the mean fitness of the population remains constant. Similar observations were made in long-term experiments with <em>Escherichia coli</em> [2], and are generally to be expected when fitness increases depend on the occurrence of beneficial mutations rather than on initial variation in the population.</p>
<p>The growth dynamics of digital organisms are comparable to those of bacteria in a chemostat. Yedid and Bell [12] investigated in detail the patterns of adaptation in digital organisms, and compared them to classic [13] and more recent [14] theories of evolving chemostat populations of bacteria. The traditional picture is that of periodic selection; that is mutants arise occasionally in an otherwise homogeneous population and either go to fixation or disappear quickly [15,16].At large population sizes or high mutation rates, however, the assumption of near homogeneity of the population cannot hold, and the concurrent existence of several beneficial mutants in a population (clonal interference) must be considered [14].Yedid and Bell varied the mutation rate over two orders of magnitude.At low mutation rates, their observations were consistent with periodic selection. Dominant genotypes were typically direct descendants of the previously dominant types, and reached abundances close to 100%. At intermediate mutation rates, this pattern started to weaken. Coexistence of advantageous mutants could be observed occasionally, and dominant abundances were generally &lt;95%. At the highest mutation rates, the populations were highly polymorphic. Dominant genotypes typically had abundances of ≤40%, and were descendants of rare, nondominant genotypes.</p>
<p>Other long-term evolutionary studies with digital organisms have addressed the growth of complexity in evolving systems [10], genome differentiation [9], or the influence of chance and history on adaptation [11]. The latter study re-created an experiment on bacteria [17] in a digital setting, with essentially similar results but higher statistical accuracy.</p>
<blockquote style="background-color: #ddddeee;">
<h1>Box 1. A brief history of digital life</h1>
<p>In the mid 1980s, an increasing number of researchers became fascinated with the idea of self-replicating computer programs. Reports of computer viruses (programs that could autonomously propagate from one computer to the next) were becoming more frequent, and certainly inspired the investigation of the ecology of reproducing computer programs in controlled environments. The earliest such investigations were in the form of games. In ‘Core War’ [a], human competitors compete against each other by writing programs in a space-limited environment (the memory of the computer). One winning strategy was writing self-replicating programs that ultimately fill up the available space with copies of itself, thus displacing the competing programs. Mutations did not occur, and the programs therefore did not evolve. Other early investigations were concerned with questions of the origin of self replication, rather than with evolution and ecology. Rasmussen et al. [b] studied the emergence of self-replicating programs in a noisy environment, but did not observe the evolution of complex programs. Tom Ray (a tropical plant ecologist by training) was the first to succeed in creating true darwinian evolution of self-replicating computer programs. In his TIERRA world (see Box Glossary) [c], self-replicating programs had to face random variations in their code, which led to a rapid diversification of the population of programs, and eventually to a significant increase in fitness. After many generations, the programs in the population were replicating much faster than were the hand-written ancestors that had seeded the initial population. However, this fitness gain was usually obtained by shrinking the program size. The evolution of complex programs from simple self replicators was first observed in the AVIDA world [d], through the evolution of computational reactions and pathways (Box 2). Finally, the AMOEBA world [e] was developed to study the emergence of self-replicating programs from nonreplicators.</p>
<h2>References</h2>
<ol type="a">
<li>Dewdney, A. (1984) <a href="/lib/mad01.html">In the game called Core War hostile programs engage in a battle of bits</a>. <em>Sci. Am.</em> 250, 14–21</li>
<li>Rasmussen, S. et al. (1990). The Coreworld: emergence and evolution of cooperative structures in a computational chemistry. <em>Physica</em> D 42, 111–134</li>
<li>Ray, T. (1991) An approach to the synthesis of life. In <em>Artificial Life II</em> (Langton, C.G. et al., eds), pp. 371-408, Addison–Wesley</li>
<li>Adami, C. (1998) <em>Introduction to Artificial Life</em>, Springer–Verlag</li>
<li>Pargellis,A.N. (1996) The spontaneous generation of digital life. <em>Physica</em> D 91, 86–96</li>
</ol>
<h2>Box Glossary</h2>
<p><strong>Amoeba:</strong> a digital life system developed by Andrew Pargellis to study the emergence of life.</p>
<p><strong>Avida:</strong> a particular software platform for digital life research developed at the California Institute of Technology. Freely available from http://dllab.caltech.edu/avida</p>
<p><strong>Tierra:</strong> the original digital life environment developed by Tom Ray. Freely available from http://www.isd.atr.co.jp/~ray/tierra/</p>
</blockquote>
<blockquote style="background-color: #ddddeee;">
<h1>Box 2. Drawbacks of the digital life approach</h1>
<p>The short generation time of digital organisms and the ease and accuracy with which measurements can be taken make digital life research appealing. However, it also has its drawbacks, both in comparison to biochemical experiments and to traditional theoretical studies based on analytic calculations or simple Monte-Carlo simulations. In contrast to purely experimental studies with bacteria or viruses, research with digital organisms is restricted to abstract questions about general principles. We cannot learn anything about the biology of a particular biochemical organism using digital life. For example, because both transcription and translation are absent in digital organisms, the evolutionary dynamics specific to these forms of expression cannot be addressed. Moreover, the design choices that enter the construction of the digital world potentially influence the outcome of experiments, so care must be taken to study only those questions for which this influence is expected to be small. Traditional theoretical or computational studies have the advantage that the particular effect under investigation can often be described more cleanly: Irrelevant details can be neglected in analytical calculations and in custom-made computer simulations, whereas the dynamics of digital organisms often become as messy as those of their biochemical counterparts. Therefore, the digital organism approach is also not the method of choice when it is relatively easy to identify which aspects of the system under study are relevant, and which can be disregarded.</p>
</blockquote>
<h2><a name="c1"></a>Epistatic interactions</h2>
<p>Understanding epistatic interactions among mutations is key to many important questions in evolutionary biology. For example, the mutational deterministic hypothesis of the origin of sex requires that the effects of several deleterious mutations are reinforcing (synergistic) in their effects [18–20], as opposed to mitigating (antagonistic). Likewise, Muller’s ratchet can operate at a significantly reduced speed in the presence of synergistic interactions, but can be accelerated by antagonistic interactions [21–23]. However, the effect of epistatic interactions on the speed of Muller’s ratchet can be offset, in part, by a continuous distribution of mutational effects of single mutations [24]. These examples underline the importance of measuring epistatic interactions as well as understanding their
 
relationship to the effects of single mutations in experimental systems. However, although there is no shortage of papers that address these issues in biochemical organisms [25–32], the corresponding experiments are very hard to perform and the results are typically of weak statistical significance [33].</p>
<p>Lenski <em>et al</em>. [34] studied the influence of genome complexity on strength and direction of epistatic interactions in digital organisms. They set up 87 strains that were adapted to an environment that was conducive to the evolution of complex COMPUTATIONAL METABOLISMS (i.e. an environment that promoted the evolution of organisms with the capability to perform complex mathematical functions; Box 3). They also derived a set of 87 strains with simple genomes by propagating each of the 87 complex strains in a simple environment in which a computational metabolism was not beneficial, and in which fitness improvements could only be achieved via a lowering of generation time. For all 174 resulting strains, they analysed the average effect of single and multiple (up to ten) mutations, with a total of between 10<sup>7</sup> and 10<sup>8</sup> separate mutants per genome. Lenski <em>et al</em>. also studied the epistatic interactions between individual pairs of mutations by comparing the joint effect of two mutations with the separate effects of each of the two mutations. The results can be summarized as follows.A statistically significant prevalence of antagonistic interactions occurred in the complex genomes, whereas, for the simple genomes, the average mutational effects of multiple mutations did not deviate significantly from the assumption of multiplicative interactions among mutations. The analysis of single pairs of mutations revealed the origin of these observations. In complex genomes, there was a substantial (19%) fraction of mutations that displayed epistatic interactions, of which antagonistic interactions were more than twice as frequent as synergistic interactions. In simple genomes, however, the antagonistic:synergistic interaction ratio was roughly equivalent to that in complex genomes, but the overall fraction of epistatic interactions between mutations was much lower (&lt;5%). Consequently, the absolute amount of excess antagonistic interactions was not large enough to bias the overall multiplicative effect of mutations. There are two conclusions to be drawn from these findings. First, an apparent lack of epistasis in the average effect of multiple mutations can in reality be caused by the cancellation of synergistic and antagonistic epistasis. Second, if these results are also representative for biochemical organisms, then the deterministic hypothesis cannot explain the origin of sex.</p>
<blockquote style="background-color: #ddddeee;">
<h1>Box 3. Computational metabolism</h1>
<p>Biochemical organisms often obtain large fitness advantages from particular metabolic pathways. For example, ATP production through respiration has a much higher yield of ATP molecules than does ATP production through fermentation, which leaves organisms possessing the respiration pathway at a selective advantage in many cases [a]. An analogous situation exists in the case of digital organisms, which execute their programs at variable speeds, thus determining their reproductive rate. The higher the speed of program execution, the faster the digital organisms reproduce. The speed of execution, in turn, is determined by the computational metabolism of an organism.</p>
<p>The computational metabolism is the set of all computational reactions that an organism performs. Computational reactions occur as follows: digital organisms can obtain numbers from their environment (these numbers can be compared to chemicals that are present in the environment of the organisms). With the right genetic code (equivalent to the sequence coding for an enzyme that catalyses a particular reaction), organisms can perform computations on these numbers. Rewarded computations are logical operations, such as bitwise AND (i.e. the logical operator between numbers) performed on the inputs. The results are then deposited back into the environment. If such a computation is deemed to be beneficial in the given environment, the organism experiences an increase in program execution speed as a result of that computation. Because different computational reactions can be strung together to create even more beneficial computations, the set of computational genes can be thought of as forming a computational pathway.</p>
<p>This computational metabolism is key to the evolution of complex organisms. In the absence of rewarded computational pathways, the only way in which organisms can increase their fitness is by shrinking their genomes as much as possible, thus minimizing the time spent copying the genome. This dynamic is then very similar to the serial transfer experiments of Spiegelman with Qβ phage [b]. When computational pathways are allowed, however, the organisms experience a tradeoff between the additional instructions spent performing calculations, which decrease the rate of replication, and the increase in program execution speed obtained from successful completion of calculations. In general, when computational pathways are rewarded with a sufficiently high acceleration, the effort of doing the computations is worth it. In that case, one can observe, over the course of several hundred generations, a tremendous increase in the complexity of these organisms, resulting eventually in organisms that can perform up to 50 or more distinct computations.</p>
<h2>References</h2>
<ol type="a">
<li>Pfeiffer, T. <em>et al</em>. (2001) Cooperation and competition in the evolution of ATP-producing pathways. <em>Science</em> 292, 504–507</li>
<li>Mills, D.R. <em>et al</em>. (1967)An extracellular Darwininan experiment with a self-duplicating nucleic acid molecule. <em>Proc. Natl. Acad. Sci. U. S. A.</em> 58, 217–224</li>
</ol>
</blockquote>
<p>The study of Lenski <em>et al</em>. was modeled after an earlier one that used <em>E. coli</em> [25]. The <em>E. coli</em> study focused on a single strain, and included only 225 mutants of that strain. There was no overall trend towards synergism or antagonism between mutations, and there were roughly equal amounts of synergistic and antagonistic interactions in the analysis of single pairs. The similarity in the results between the <em>E. coli</em> and digital organisms studies is striking, and supports the hypothesis that many aspects of evolving systems are governed by general principles.</p>
<p>A type of universality in evolving systems also emerges from an investigation of the correlation between the average effect of single mutations and the epistatic interactions among multiple mutations [35] that used both the digital genomes obtained by Lenski <em>et al</em>. [34] and newly derived data from RNA secondary structure folding. For both data sets, the average effect of single mutations was positively correlated to the amount and strength of antagonistic interactions; that is, genomes that were strongly affected by single mutations showed an elevated level of antagonistic interactions and vice versa. This observation is the result of a geometrical constraint on the distribution of viable sequences in sequence space [35–37], and is therefore expected to hold universally.</p>
<h2><a name="c2"></a>Quasi-species dynamics</h2>
<p>The relevance of quasi-species evolution [38] for the understanding of bacterial and viral dynamics has
 
been debated for the past 20 years [39,40]. In a nutshell, the quasi-species concept states that asexual organisms evolve as cohesive groups of closely related mutants, and that selection acts on these mutant clouds (quasi-species), rather than on the individual organisms. To observe quasi-species effects, the mutation rates must be relatively high, of the order of one mutation per genome and generation, which disqualifies most bacterial systems as models in which evidence of quasi-species evolution could be observed. The situation is different in RNA viruses, which evolve at such high mutation rates [41]. Likewise, digital organisms are expected to behave as predicted by the quasi-species model at correspondingly high mutation rates [42].</p>
<p>One of the most spectacular predictions of quasi-species theory is that a fast replicator will be less ‘fit’than a slower replicator with better mutational support. In other words, a mutational cloud that contains some very fast replicating individuals and some slowly replicating ones can be out-competed by another mutational cloud in which most of the individuals replicate at a medium speed, and fast replicators are absent. The theory behind this effect has been known since 1988 [43], but had negligible impact on virology research. Recently, it has been possible to demonstrate this effect in digital organisms [44]. Wilke et al. propagated 40 evolved strains of organisms for 1000 generations in both low and high mutation rate environments. This resulted in 40 pairs of organisms, each pair comprising a strain evolved at a low mutation rate (the ‘A’strain) and a strain evolved at high mutation rate (the ‘B’strain). Both strains shared a common ancestor 1000 generations in the past. Out of the 40 pairs, 36 had anA strain that had evolved a higher replication rate than had the B strain. In 12 of these cases, the replication rate of A exceeded that of B by >50%. TheA and B strains of these 12 pairs subsequently competed with each other in environments with different mutation rates. Without exception, the competition experiments were won byA strains at low mutation rates and by B strains at high mutation rates, notwithstanding their replicative disadvantage. The crucial mutation rate at which the experimental outcome would switch in B’s favor could be predicted accurately from quasi-species theory. Such experiments imply that mutational robustness plays an increasing role in the expectation of long-term evolutionary success as the mutation rate increases. Because of the general nature of the results, it is expected that the same effect can be observed with biochemical organisms as long as the relevant experimental conditions can be created.</p>
<h2><a name="c3"></a>Future research directions</h2>
<h3><a name="c31"></a>Digital life genetics</h3>
<p>To a human eye, the genome of an evolved digital organism appears to be a random collection of computer instructions, assembled without any planning or organization. However, a detailed inspection reveals that these genomes are surprisingly well organized, and that they can often be subdivided into functionally distinct blocks, which deserve to be called genes. These genes can be discovered as follows: one systematically replaces each instruction of the genome, one at a time, with a special null instruction that has no function. Then, one tests each of the organisms obtained in that manner for their ability to replicate, for their speed of replication, and for the COMPUTATIONAL REACTIONS they can complete. In that way, one obtains a map of the parts of the genome that are essential for replication, the ones that only play a role in certain computational reactions but are not vital, and the ones that have no discernible function (junk genes). This method of mapping out genomes can then be applied to closely related mutants, for example to a sequence of successive descendants taken from an evolving population. The mechanisms by which evolution proceeds will be revealed in detail in such a study, and it will be possible to identify, for instance, whether new computational reactions form de novo out of junk genes, or rather arise from genes that code for other, preexisting reactions.At the time of writing, no such study has been published, although the necessary tools are available.</p>
<h3><a name="c32"></a>Evolutionary ecology</h3>
<p>Work on digital organisms to date has focused on single-niche systems, in which the organisms interact only indirectly through their difference in speed of replication. The computational metabolisms are not affected by the presence of competitors that perform either similar or different computations, implying an absence of frequency-dependent selection. An obvious direction for future work is to create ecological interactions by coupling the efficiency of the computational metabolism of the organisms to the presence or absence of external resources [45], in the following manner: to each computation, assign a necessary resource.An organism performing a computation (a mathematical reaction) increases its execution speed by depleting the associated resource. If the resource is abundant, the organism can reap large benefits, whereas if the resource is scarce, that particular reaction is not advantageous to the organism. Moreover, it is possible to add resource conversion, so that when a certain resource is depleted on completion of a particular reaction, another one is incremented. These resource dynamics, perhaps combined with restrictions of reactions (e.g. organisms performing reaction A cannot perform reaction B, or organisms can perform reaction B only after they have already completed reactionA), should lead to multi-niche systems with rich ecological interactions among digital organisms, including frequency-dependent selection.</p>
<h2><a name="c4"></a>Acknowledgments</h2>
<p>We thank R.E. Lenski and C. Ofria for extensive discussions and encouragement. This work was supported by the National Science Foundation under contract No DEB-9981397. Part of this work was carried out at the Jet Propulsion Laboratory, under a contract with the National Aeronautics and Space Administration.</p>
<h2><a name="c5"></a>Conclusions</h2>
<p>Research on digital life forms has reached a level of sophistication at which questions of biological
 
relevance can be both addressed and answered. The main focus of digital life research is a sort of comparative biology, which attempts to extricate those aspects of simple living systems that are germane to the type of chemistry used, from those that are not [46].Additionally, digital life can help to refine mathematical theories and aid in developing and quickly testing new hypotheses about ecological and evolutionary processes.</p>
<p>Current state-of-the-art digital life research platforms create essentially single-niche ecosystems, and the dynamics that unfold are similar to bacterial or viral evolution in chemostats. However, more complex ecosystem structures can be realized within the paradigm of self-replicating computer programs, and we can expect research to increase in this area.</p>
<p>Apart from their usefulness as a tool of understanding evolution in general, it is important to study the biology of artificial life forms in their own right. Recently, Lipson and Pollack showed that the principles of simple self-replicating robots are within reach of current technology [47]. Eventually, such robots, and the software that directs them, might evolve without human interaction, at which point they would become part of the ecosystem in which we live.</p>
<h2><a name="c6"></a>Glossary</h2>
<p><strong>Computational metabolism:</strong> the total set of computational reactions that an organism can do.</p>
<p><strong>Computational reaction:</strong> a computation carried out by an organism on numbers provided by the environment. To achieve such a reaction, the organisms must possess a computational gene (code) or pathway for this reaction.</p>
<p><strong>Explicit mutation:</strong> change in the genome between parent and offspring that is caused by noise in the environment (copy mutation, insertion mutation, deletion mutation, etc.)</p>
<p><strong>Genome:</strong> the program (comprised of instructions arranged in a circular fashion) that an organism executes.</p>
<p><strong>Instruction:</strong> the basic unit of information in the genome of a digital organism. Each instruction corresponds typically to one unique action executed by the central processing unit. For example, the copy instruction copies a single instruction in the genome of the organism, and the divide instruction initiates the separation of the daughter genome after copying the genome line by line.</p>
<p><strong>Implicit mutation:</strong> change in the genome between parent and offspring that is caused by a malfunctioning copy process of the parent organism.</p>
<h2><a name="c7"></a>References</h2>
<ol>
<li>Foster, J.A. (2001) Evolutionary computation. <em>Nat. Rev. Genet.</em> 2, 428–436</li>
<li>Lenski, R.E. and Travisano, M. (1994) Dynamics of adaptation and diversification: a 10,000-generation experiment with bacterial populations. <em>Proc. Natl. Acad. Sci. U. S. A.</em> 91, 6808–6814</li>
<li>Sniegowski, P.D. <em>et al</em>. (1997) Evolution of high mutation rates in experimental populations of <em>E. coli</em>. <em>Nature</em> 387, 703–705</li>
<li>Cooper, V.S. and Lenski, R.E. (2000) The population genetics of ecological specialization in evolving <em>E. coli</em> populations. <em>Nature</em> 407, 736–739</li>
<li>Ray, T. (1991) An approach to the synthesis of life. In <em>Artificial Life II</em> (Langton, C.G. <em>et al.</em>, eds), pp. 371-408, Addison–Wesley</li>
<li>Ray, T.S. (1994) Evolution, complexity, entropy, and artificial reality. <em>Physica</em> D 75, 239–263</li>
<li>Adami, C. (1995) Self-organized criticality in living systems. <em>Phys. Lett.</em> A 203, 29–32</li>
<li>Adami, C. (1998) <em>Introduction to Artificial Life</em>, Springer–Verlag</li>
<li>Ofria, C. <em>et al.</em> (1999) Evolution of differentiated expression patterns in digital organisms. <em>Lect. Notes Artif. Intell.</em> 1674, 129–138</li>
<li>Adami, C. <em>et al.</em> (2000) Evolution of biological complexity. <em>Proc. Natl. Acad. Sci. U.S.A.</em> 97, 4463–4468</li>
<li>Wagenaar, D. and Adami, C. (2000) Influence of chance, history and adaptation on evolution in digitalia. In <em>Artificial Life VII</em> (Bedau, M.A. <em>et al.</em>, eds), pp. 216–220, MIT Press</li>
<li>Yedid, G. and Bell, G. (2001) Microevolution in an electronic microcosm. <em>Am. Nat.</em> 157, 465–487</li>
<li>Atwood, K.C. <em>et al.</em> (1951) Selective mechanisms in bacteria. <em>Cold Spring Harbor Symp. Quant. Biol.</em> 16, 146–155</li>
<li>Gerrish, P.J. and Lenski, R.E. (1998) The fate of competing beneficial mutations in an asexual population. <em>Genetica</em> 103, 127–144</li>
<li>Notley-McRobb, L. and Ferenci, T. (2000) Experimental analysis of molecular events during mutational periodic selections in bacterial evolution. <em>Genetics</em> 156, 1493–1501</li>
<li>Imhof, M. and Schlotterer, C. (2001) Fitness effects of advantageous mutations in evolving <em>Escherichia coli</em> populations. <em>Proc. Natl. Acad. Sci. U.S.A.</em> 98, 1113–1117</li>
<li>Travisano, M. et al. (1995) Experimental test of the roles of adaptation, chance, and history in evolution. <em>Science</em> 267, 87–90</li>
<li>Kondrashov, A.S. (1982) Selection against harmful mutations in large sexual and asexual populations. <em>Genet. Res.</em> 40, 325–332</li>
<li>Kondrashov, A.S. (1988) Deleterious mutations and the evolution of sexual reproduction. <em>Nature</em> 336, 435–440</li>
<li>West, S.A. <em>et al.</em> (1999) A pluralist approach to sex and recombination. <em>J. Evol. Biol.</em> 12, 1003–1012</li>
<li>Crow, J.F. and Kimura, M. (1978) Efficiency of truncation selection. <em>Proc. Natl. Acad. Sci. U. S. A.</em> 76, 396–399</li>
<li>Kondrashov, A.S. (1994) Muller’s ratchet under epistatic selection. <em>Genetics</em> 136, 1469–1473</li>
<li>Colato, A. and Fontanari, J.F. (2001) Soluble model for the accumulation of mutations in asexual populations. <em>Phys. Rev. Lett.</em> 87, 238102</li>
<li>Butcher, D. (1995) A general model for the evolution of recombination. <em>Genet. Res.</em> 65, 123–144</li>
<li>Elena, S.F. and Lenski, R.E. (1997) Test of synergistic interactions among deleterious mutations in bacteria. <em>Nature</em> 390, 395–398</li>
<li>De Visser, J.A.G.M. <em>et al.</em> (1997) An experimental test for synergistic epistasis and its application to Chlamydomonas. <em>Genetics</em> 145, 815–819</li>
<li>De Visser, J.A.G.M. and Hoekstra, R. F. (1998) Synergistic epistasis between loci affecting fitness: evidence in plants and fungi. <em>Genet. Res.</em> 71, 39–49</li>
<li>Elena, S.F. (1999) Little evidence for synergism among deleterious mutations in a nonsegmented RNA virus. <em>J. Mol. Evol.</em> 49, 703–707</li>
<li>de la Peña, M. <em>et al.</em> (2000) Effect of deleterious mutation accumulation on the fitness of RNA bacteriophage MS2. <em>Evolution</em> 54, 686–691</li>
<li>Whitlock, M.C. and Bourguet, D. (2000) Factors affecting the genetic load in <em>Drosophila</em>: synergistic epistasis and correlations among fitness components. <em>Evolution</em> 54, 1654–1660</li>
<li>Peters, A.D. and Keightley, P.D. (2000) A test for epistasis among induced mutations in <em>Caenorhabditis elegans. Genetics</em> 156, 1635–1647</li>
<li>Wloch, D.M. et al. (2001) Epistatic interactions of spontaneous mutations in haploid strains of the yeast <em>Saccharomyces cerevisiae</em>. <em>J. Evol. Biol.</em> 14, 310–316</li>
<li>West, S.A. <em>et al.</em> (1998) Testing for epistasis between deleterious mutations. <em>Genetics</em> 149, 435–444</li>
<li>Lenski, R.E. <em>et al.</em> (1999) Genome complexity, robustness, and genetic interactions in digital organisms. <em>Nature</em> 400, 661–664</li>
<li>Wilke, C.O. andAdami, C. (2001) Interaction between directional epistasis and average mutational effects. <em>Proc. R. Soc. Lond. Ser. B</em> 268, 1469–1474</li>
<li>Wagner, G.P. <em>et al.</em> (1998) Genetic measurement theory of epistatic effects. <em>Genetica</em> 102/103, 569–580</li>
<li>Rice, S.H. (1998) The evolution of canalization and the breaking of von Baer’s law: modeling the evolution of development with epistasis. <em>Evolution</em> 52, 647–656</li>
<li>Eigen, M. and Schuster, P. (1979) <em>The Hypercycle – A Principle of Natural Self-Organization</em>, Springer–Verlag</li>
<li>Holmes, E.C. and, Moya A. (2002) Is the quasi-species concept relevant to RNA viruses? <em>J. Virol.</em> 76, 460–462</li>
<li>Domingo, E. (2002) Quasi-species theory in virology. <em>J. Virol.</em> 76, 463–465</li>
<li>Burch, C.L. and Chao, L. (2000) Evolvability of an RNA virus is determined by its mutational neighbourhood. <em>Nature</em> 406, 625–628</li>
<li>Wilke, C.O. (2002) Maternal effects in molecular evolution. <em>Phys. Rev. Lett.</em> 88, 078101</li>
<li>Schuster, P. and Swetina, J. (1988) Stationary mutant distributions and evolutionary optimization. <em>Bull. Math. Biol.</em> 50, 635–660</li>
<li>Wilke, C.O. <em>et al.</em> (2001) Evolution of digital organisms at high mutation rate leads to survival of the flattest. <em>Nature</em> 412, 331–333</li>
<li>Adami, C. (2002) <em>Ab initio</em> modeling of ecosystems with artificial life. <em>Nat. Res. Mod.</em> 15, 133–145</li>
<li>Maynard Smith, J. (1992) Byte-sized evolution. <em>Nature</em> 335, 772–773</li>
<li>Lipson, H. and Pollack, J.B. (2000) Automatic design and manufacture of robotic lifeforms. <em>Nature</em> 406, 974–978</li>
</ol>
 
[<a style="" href="/lib/?lang=EN&amp;index=AI#mcw00">Back to index</a>] [<a href="/lib/mcw00.html#disqus_thread">Comments</a>]<br/> <div id="disqus_thread"></div>
<script type="text/rocketscript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'vxheaven'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
<noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript></div>
<div><small>By accessing, viewing, downloading or otherwise using this content you agree to be bound by the <a href="/agreement.php">Terms of Use</a>!</small> <small>vxheaven.org aka vx.netlux.org</small></div>
<div style="margin-top: 2px; float: left;" class="adsapeu">
<script type="text/rocketscript">
<!--
var _acic={dataProvider:10};(function(){var e=document.createElement("script");e.type="text/javascript";e.async=true;e.src="//www.acint.net/aci.js";var t=document.getElementsByTagName("script")[0];t.parentNode.insertBefore(e,t)})()
//-->
</script>
</div>
<script data-rocketsrc="http://www.google-analytics.com/urchin.js" type="text/rocketscript"></script><script type="text/rocketscript">try { _uacct = "UA-590608-1"; urchinTracker(); } catch(err) {}</script>
<div style="display: none;"><a href="/lib/index.php?lang=de&amp;id=mcw00">de</a><a href="/lib/index.php?lang=en&amp;id=mcw00">en</a><a href="/lib/index.php?lang=es&amp;id=mcw00">es</a><a href="/lib/index.php?lang=it&amp;id=mcw00">it</a><a href="/lib/index.php?lang=fr&amp;id=mcw00">fr</a><a href="/lib/index.php?lang=pl&amp;id=mcw00">pl</a><a href="/lib/index.php?lang=ru&amp;id=mcw00">ru</a><a href="/lib/index.php?lang=ua&amp;id=mcw00">ua</a></div>
</body>
</html>
