<?xml version="1.0"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<title> SPTH 'Imitation of Life: Advanced system for native Artificial Evolution' (VX heaven)</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8"/>
<meta name="Author" content="SPTH"/>
<meta name="KeyWords" lang="en" content="computer virus, virus, virii,vx, компьютерные вирусы, вирус, вири, SPTH,Imitation of Life: Advanced system for native Artificial Evolution, rega, energy, instruction, rate, hamming, distance, mutation, save, instructions, codon, system, process, removed, thomas, offset"/>
<meta name="Description" content="A model for artificial evolution in native x86 Windows systems has been developed at the end of 2010. In this text, further improvements and additional analogies to natural microbiologic processes are presented. Several experiments indicate the capability of the system - and raise the question of possible countermeasures."/>
<script type="text/javascript">
//<![CDATA[
try{if (!window.CloudFlare) {var CloudFlare=[{verbose:0,p:0,byc:0,owlid:"cf",bag2:1,mirage2:0,oracle:0,paths:{cloudflare:"/cdn-cgi/nexp/dok3v=1613a3a185/"},atok:"047a5bcbf67431883fc9ed25fba33612",petok:"83543a3142b81cd83d0a98070eae8e08ca44a08d-1498756461-1800",zone:"vxheaven.org",rocket:"a",apps:{}}];document.write('<script type="text/javascript" src="//ajax.cloudflare.com/cdn-cgi/nexp/dok3v=85b614c0f6/cloudflare.min.js"><'+'\/script>');}}catch(e){};
//]]>
</script>
<link rel="icon" href="/favicon.ico" type="image/x-icon"/>
<link rel="shortcut icon" href="/favicon.ico" type="image/x-icon"/>
<link rel="stylesheet" type="text/css" href="/style.css"/><link rel="canonical" href="http://vxheaven.org/lib/vsp37.html"/>
<script type="text/rocketscript" data-rocketsrc="https://apis.google.com/js/plusone.js">{"parsetags": "explicit"}</script>
</head>
<body bgcolor="#dbc8a0" text="#302000" link="#225599" vlink="#113366">
<div class="s1">
<div style="float:right;"><a href="/lib/index.php?tbs=1"><img src="/img/max.gif" alt="Maximize"/></a></div> <form id="lf" style="margin: 0; float: right;" method="get" action="/index.php"><input type="hidden" name="action" value="set"/><select name="lang" onchange="javascript:document.getElementById('lf').submit();"><option value="ru">Русский</option><option selected="selected" value="en">English</option><option value="ua">Українська</option><option value="de">Deutsch</option><option value="es">Español</option><option value="fr">Fran&ccedil;ais</option><option value="it">Italiano</option><option value="pl">Polski</option></select></form>
<div style="float: right;"><div id="plusone"></div></div>
<script type="text/rocketscript">gapi.plusone.render("plusone", {"size":"small","count":"true"});</script>
<div style="float: right;" class="addthis_toolbox addthis_default_style">
<script type="text/rocketscript">var addthis_config = { ui_click: true }</script>
<a style="text-decoration: none; font-size: 10pt;" href="/?action=addthis" class="addthis_button_compact">Bookmark</a>
<script type="text/rocketscript" data-rocketsrc="http://s7.addthis.com/js/250/addthis_widget.js#username=herm1t"></script>
</div>
<div style="float: right;">
<script type="text/rocketscript" data-rocketsrc="http://www.google.com/cse/brand?form=cse-search-box&amp;lang=en"></script>
<form action="/search.php" id="cse-search-box">
<input type="hidden" name="cx" value="002577580816726040001:z9_irkorydo"/>
<input type="hidden" name="cof" value="FORID:10"/>
<input type="hidden" name="ie" value="UTF-8"/>
<input type="text" name="q" size="12" value=" "/>
<input type="submit" name="sa" value="Search"/>
</form>
</div><h1><a href="/" style="text-decoration: none; color: #000000;">VX Heaven</a></h1>
<span class="nav"><a href="/lib/">Library</a> <a href="/vl.php">Collection</a> <a href="/src.php">Sources</a> <a href="/vx.php?id=eidx">Engines</a> <a href="/vx.php?id=tidx">Constructors</a> <a href="/vx.php?id=sidx">Simulators</a> <a href="/vx.php?id=uidx">Utilities</a> <a href="/links.php">Links</a> <a href="/donate.php" style="color: #706020" id="donate">Donate</a> <a href="/forum" style="text-decoration: underline;">Forum</a> </span><br clear="all"/>
</div>
<div class="s2"><h1>Imitation of Life: Advanced system for native Artificial Evolution</h1><p><a href="/lib/?lang=en&amp;author=SPTH"> SPTH</a><br/> <em><a href="/vx.php?fid=1943#f1943">Valhalla #1</a></em><br/> <em>July 2011</em></p><script type="text/rocketscript">var disqus_url = 'http://vxheaven.org/lib/vsp37.html';</script><div class="ci"><a href="/lib/?ci=vsp37">1</a></div><img src="/img/pdf.gif" alt="PDF"/><a href="/lib/pdf/Imitation%20of%20Life%3A%20Advanced%20system%20for%20native%20Artificial%20Evolution.pdf">Download</a> PDF (624.63Kb) (You need to be registered on <a href="/forum">forum</a>)<br/>[<a style="" href="/lib/?lang=EN&amp;index=AI#vsp37">Back to index</a>] [<a href="/lib/vsp37.html#disqus_thread">Comments</a>]<br/> <form method="post" action="">
<img src="/img/cache/0b9fd596a90421f9f1f68a9760275737.gif" alt="\text{T_EX size}" valign="middle"/>
<select name="TeX_size"><option value="-2">-2</option><option value="-1">-1</option><option value="0" selected="selected">0</option><option value="1">1</option><option value="2">2</option><option value="3">3</option><option value="4">4</option><option value="5">5</option> </select>
<input type="submit" value="Scale"/>
</form>
<address>
<a class="__cf_email__" href="/cdn-cgi/l/email-protection" data-cfemail="2a595a4f5846045e4245474b596a4d474b434604494547">[email&#160;protected]</a><script data-cfhash='f9e31' type="text/rocketscript">/* <![CDATA[ */!function(t,e,r,n,c,a,p){try{t=document.currentScript||function(){for(t=document.getElementsByTagName('script'),e=t.length;e--;)if(t[e].getAttribute('data-cfhash'))return t[e]}();if(t&&(c=t.previousSibling)){p=t.parentNode;if(a=c.getAttribute('data-cfemail')){for(e='',r='0x'+a.substr(0,2)|0,n=2;a.length-n;n+=2)e+='%'+('0'+('0x'+a.substr(n,2)^r).toString(16)).slice(-2);p.replaceChild(document.createTextNode(decodeURIComponent(e)),c)}p.removeChild(t)}}catch(u){}}()/* ]]> */</script><br/>
<a href="http://www.spth.de.vu/">www.spth.de.vu</a>
</address>
<ul>
<li><a href="#c0">Abstract</a></li>
<li><a href="#c1">1 Introduction</a></li>
<li><a href="#c2">2 Artificial Biosynthesis</a>
<ul>
<li><a href="#c21">2.1 Meta-Language</a>
<ul>
<li><a href="#c211">2.1.1 Mutable API calls</a></li>
<li><a href="#c212">2.1.2 Example: ROL instruction</a></li>
</ul></li>
<li><a href="#c22">2.2 Translator</a></li>
<li><a href="#c23">2.3 Replication, Mutation and Selection</a></li>
</ul></li>
<li><a href="#c3">3 Mutations</a>
<ul>
<li><a href="#c31">3.1 Point mutation - bit flip</a></li>
<li><a href="#c32">3.2 Chromosomal inversion - byte eXCHanGe</a></li>
<li><a href="#c33">3.3 Deletion, insertion, translocation</a></li>
<li><a href="#c34">3.4 Horizontal gene transfer</a></li>
<li><a href="#c35">3.5 Polymorphism: neutral codon variation</a></li>
</ul></li>
<li><a href="#c4">4 Further improvements</a>
<ul>
<li><a href="#c41">4.1 Start- and Stop codons: Splicing</a></li>
<li><a href="#c42">4.2 Optimization of the instruction set</a></li>
<li><a href="#c43">4.3 Optimization of the alphabet</a></li>
</ul></li>
<li><a href="#c5">5 Experiments</a>
<ul>
<li><a href="#c51">5.1 Hamming distance</a></li>
<li><a href="#c52">5.2 Effect of alphabet "energy"</a></li>
<li><a href="#c53">5.3 Effect of Start- and Stop-codons</a></li>
</ul></li>
<li><a href="#c6">6 Outlook</a></li>
<li><a href="#c7">7 References</a></li>
</ul>
<h2><a name="c0"></a>Abstract</h2>
<p>A model for artificial evolution in native x86 Windows systems has been developed at the end of 2010. In this text, further improvements and additional analogies to natural microbiologic processes are presented. Several experiments indicate the capability of the system - and raise the question of possible countermeasures.</p>
<h2><a name="c1"></a>1 Introduction</h2>
<p>Artificial Evolution has become a successful playground for evolutional experiments, when Tom Ray released the Tierra system[<a href="#f1_1" name="b1_1">1</a>]. Tierra is a virtual system with self-replicating programs which simulate mutations in form of copying errors. The artificial creatures struggle for the limited resources (such as CPU time and memory space), thus the systems fullfills the three criteria for evolution: replication, mutation, selection.</p>
<p>In order to achieve high robustness against mutations, Ray introduced techniques such as non-direct addressing and seperation of arguments and operations.</p>
<p>Many interesting insights to evolution have been found with Tierra (such as evolution of multi-cellularity[<a href="#f2_2" name="b2_2">2</a>] or parallel computing[<a href="#f3_3" name="b3_3">3</a>]) and similar systems such as avida (evolution under high mutation rate[<a href="#f4_4" name="b4_4">4</a>] and emergence of complex features[<a href="#f5_5" name="b5_5">5</a>]).</p>
<p>Iliopoulos, Adami and Ször have discussed the consequences for computer security of implementing darwinian principles into native system, in 2008.[<a href="#f7_7" name="b7_7">7</a>] They concluded that a truly undetectable virus might be more feasible than previously imagined, and that it is currently unknown whether there would be a defence against such organisms.</p>
<p>In 2010, I have created the first (to my knowledge) implementation of an artificial evolution system for a native operation system (Microsoft Windows XP+ 32bit), using several parallels to the natural biosynthesis process[<a href="#f8_8" name="b8_8">8</a>]. A short comparison between usual x86 code and the new artificial evolution concept shows that the new concept is actually more robust against mutations.[<a href="#f9_9" name="b9_9">9</a>]</p>
<p>The presented proof-of-concept organism as well as the underlaying metalanguage have been analysed in detail by Peter Ferrie in 2011[<a href="#f10_10" name="b10_10">10</a>][<a href="#f11_11" name="b11_11">11</a>].</p>
<p>The idea of this article is to continue this research...</p>
<h2><a name="c2"></a>2 Artificial Biosynthesis</h2>
<p>The main idea is to use a similar concept to natural biosynthesis: The codons of the mRNA are translated into amino acids using tRNA molecules, these amino acid chains form the proteins - the actually functional part of the cell.</p>
<p>In the artificial analogon, the whole information of the organism is saved in a chain of codons. Each codon consists of 8bits, thus there are 256 different codons. In the translator (similar to tRNA in the ribosom), the codons will be mapped to an x86 instruction (similar to an amino acid) - a chain of x86 instructions form the protein, the functional part of the organism:</p>
<table summary="Artificial vs. Natural">
<tr><th>Artificial</th><th>Natural</th></tr>
<tr><td>bit</td><td>nucleobase</td></tr>
<tr><td>byte</td><td>codon</td></tr>
<tr><td>instruction</td><td>amino acid</td></tr>
<tr><td>function</td><td>protein</td></tr>
<tr><td>translator</td><td>tRNA</td></tr>
</table>
<p>In the natural system, the information is saved in codons, which consists of 3 nucleobases - as there exists 4 different nucleobases (<strong>A</strong>denine, <strong>G</strong>uanine, <strong>T</strong>hymine, <strong>C</strong>ytosine), a codon can have 4<sup>3</sup> = 64 different representations. Each codon codes one out of 20 amino acids, thus there is a redundancy in the mapping process, which is used to increase the robustness of the code. This redundancy is also used in the artificial biosynthesis as there are less than 2<sup>8</sup>=256 base functions of the meta language.</p>
<h3><a name="c21"></a>2.1 Meta-Language</h3>
<p>The idea is to create a compact, <em>complete</em> instruction set with seperation of arguments and operations, and with non-direct addressing.</p>
<p>The language provides seven registers with specific properties:</p>
<ul>
<li><strong>RegA, RegB, RegD</strong> - general purpose registers (correspond to EAX, EBX, EDX)</li>
<li><strong>BC1</strong> - operation register (correspond to EBX): the first argument of every operation, and source or destination for other instructions</li>
<li><strong>BC2</strong> - argument register (correspond to ECX): the second argument of every operation</li>
<li><strong>BA1</strong> - write address register (correspond to EDI): holds the address for write instructions</li>
<li><strong>BA2</strong> - jump address register (correspond to ESI): holds the address for jump instructions</li>
</ul>
<p>The seperation of arguments and operations is realized by using BC1 (and BC2) as standard arguments, and filling them independently of the operation.</p>
<p>The language provides PIC (<em>position-independent code</em>). Every address is relative to the instruction pointer thus is independent of the position.</p>
<table summary="Instruction set">
<tr><th>Instruction</th><th>HLL</th><th>Assembler</th></tr>
<tr><td>nopREAL</td><td> </td><td>nop</td></tr>
<tr><td>nopsA</td><td>BC1 = RegA</td><td>mov ebx, eax</td></tr>
<tr><td>nopsB</td><td>BC1 = RegB</td><td>mov ebx, ebp</td></tr>
<tr><td>nopsD</td><td>BC1 = RegD</td><td>mov ebx, edx</td></tr>
<tr><td>nopdA</td><td>RegA = BC1</td><td>mov eax, ebx</td></tr>
<tr><td>nopdB</td><td>RegB = BC1</td><td>mov ebp, ebx</td></tr>
<tr><td>nopdD</td><td>RegD = BC1</td><td>mov edx, ebx</td></tr>
<tr><td>save</td><td>BC2 = BC1</td><td>mov ecx, ebx</td></tr>
<tr><td>addsaved</td><td>BC1 + = BC2</td><td>add ebx, ecx</td></tr>
<tr><td>subsaved</td><td>BC1 - = BC2</td><td>sub ebx, ecx</td></tr>
<tr><td>saveWrtOff</td><td>BA1 = BC1</td><td>mov edi, ebx</td></tr>
<tr><td>saveJmpOff</td><td>BA2 = BC1</td><td>mov esi, ebx</td></tr>
<tr><td>writeByte</td><td>byte[BA1] = (BC1 &amp; 0xFF)</td><td>mov byte[edi], bl</td></tr>
<tr><td>writeDWord</td><td>dword[BA1] = BC1</td><td>mov dword[edi], ebx</td></tr>
<tr><td>getDO</td><td>BC1 = DataOffset</td><td>mov ebx, DataOffset</td></tr>
<tr><td>getdata</td><td>BC1 = dword[BC1]</td><td>mov ebx, dword[ebx]</td></tr>
<tr><td>getEIP</td><td>BC1 = InstructionPointer</td><td>call gEIP; gEIP: pop ebx</td></tr>
<tr><td>push</td><td>push BC1</td><td>push ebx</td></tr>
<tr><td>pop</td><td>pop BC1</td><td>pop ebx</td></tr>
<tr><td>pushall</td><td>pushad</td><td>pushad</td></tr>
<tr><td>popall</td><td>popad</td><td>popad</td></tr>
<tr><td>zer0</td><td>BC1 = 0</td><td>mov ebx, 0x0</td></tr>
<tr><td>add0001</td><td>BC1 + = 0x1</td><td>add ebx, 0x1</td></tr>
<tr><td>add0004</td><td>BC1 + = 0x4</td><td>add ebx, 0x4</td></tr>
<tr><td>add0010</td><td>BC1 + = 0x10</td><td>add ebx, 0x10</td></tr>
<tr><td>add0040</td><td>BC1 + = 0x40</td><td>add ebx, 0x40</td></tr>
<tr><td>add0100</td><td>BC1 + = 0x100</td><td>add ebx, 0x100</td></tr>
<tr><td>add0400</td><td>BC1 + = 0x400</td><td>add ebx, 0x400</td></tr>
<tr><td>add1000</td><td>BC1 + = 0x1000</td><td>add ebx, 0x1000</td></tr>
<tr><td>add4000</td><td>BC1 + = 0x4000</td><td>add ebx, 0x4000</td></tr>
<tr><td>sub0001</td><td>BC1 - = 0x1</td><td>sub ebx, 0x1</td></tr>
<tr><td>shl</td><td>BC1 &lt;&lt; (BC2 &amp; 0xFF)</td><td>shl ebx, cl</td></tr>
<tr><td>shr</td><td>BC1 >> (BC2 &amp; 0xFF)</td><td>shr ebx, cl</td></tr>
<tr><td>xor</td><td>BC1 ^ = BC2</td><td>xor ebx, ecx</td></tr>
<tr><td>and</td><td>BC1 &amp; = BC2</td><td>and ebx, ecx</td></tr>
<tr><td>mul</td><td>(RegD: RegA) = RegA * BC1</td><td>mul ebx</td></tr>
<tr><td>div</td><td>(RegD, RegA) = RegA / BC1</td><td>div ebx</td></tr>
<tr><td>JnzUp</td><td>&nbsp;</td><td>jz over; jmp esi; over:</td></tr>
<tr><td>JnzDown</td><td>&nbsp;</td><td>jnz down; times 32: nop; down:</td></tr>
<tr><td>call</td><td>&nbsp;</td><td>stdcall ebx</td></tr>
<tr><td>CallAPILoadLibrary</td><td>&nbsp;</td><td>stdcall dword[LoadLibrary]</td></tr>
</table>
<h4><a name="c211"></a>2.1.1 Mutable API calls</h4>
<p>To use the APIs provided by the OS, an organism can use the call instruction:</p>
<div class="asm" style="font-family:monospace;color: #000066;  border: solid thin #c2c1b1; background: #d6d5c5;">&nbsp; DataOffset<span style="color: #339933;">:</span><br/>
&nbsp; <span style="color: black; font-style: italic;">; [...]</span><br/>
&nbsp; &nbsp; &nbsp; hGetCommandLineA <span style="color: #0000ff; font-weight: bold;">dd</span> <span style="color: #ff0000;">0x0</span> &nbsp; &nbsp;<span style="color: black; font-style: italic;">; should contain</span><br/>
&nbsp; <span style="color: black; font-style: italic;">; [...] &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;; addresse of API</span><br/>
<br/>
&nbsp; <span style="color: #339933;">.</span><span style="color: #0000ff; font-weight: bold;">code</span><br/>
&nbsp; <span style="color: #0000ff; font-weight: bold;">start</span><span style="color: #339933;">:</span><br/>
&nbsp; <span style="color: black; font-style: italic;">; [...]</span><br/>
&nbsp; &nbsp; &nbsp; _getDO &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; BC1=DataOffset</span><br/>
&nbsp; &nbsp; &nbsp; addnumber <span style="color: black;">&#40;</span>hGetCommandLineA<span style="color: #339933;">-</span>DataOffset<span style="color: black;">&#41;</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; macro to add number</span><br/>
&nbsp; &nbsp; &nbsp; _getdata &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; BC1=dword[hGetCommandLineA]</span><br/>
<br/>
&nbsp; &nbsp; &nbsp; _call &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; call BC1 :)</span><br/>
&nbsp; <span style="color: black; font-style: italic;">; [...]</span><br/>
&nbsp; <span style="color: #339933;">.</span>end <span style="color: #0000ff; font-weight: bold;">start</span><br/>
&nbsp;</div>
<p>One possibility is to save the API addresses directly within the organism body. However, this has two disadvantages: Firstly, the addresses may change for each new version of Windows. Secondly, the probability that a different and valid API address appears out of one single bitflip is very low (a rough estimate: a memory address of 32-bit gives 2^32 possible addresses - say there are 5000 valid API addresses. The probability to reach one of them is P = 5000/2^32 ~= 10^-6).</p>
<p>A different method is to save a short hash of the desired API name in the organism. Then load the DLL, scan the export section for API names and create hashes for each API. If the hashes match, save the address of the API. This approach is independent of the OS version, and it's very mutable. It is possible to use hash as short as 12bit for each API. Let's say there are 1000 APIs in a DLL file. The probability to access on of these APIs within one single bit ip is given by P = 1000/2^12 ~= 0.25. In average, every 4th bitflip in the API hash leads to a different hash corresponding to a valid API.</p>
<h4><a name="c212"></a>2.1.2 Example: ROL instruction</h4>
<p>The x86 instruction ROL (<em>Rotate Left</em>) is not directly provided by the instruction set. However, it can be written using just instructions of the set.</p>
<p>Let's say, one would like to write <em>rol RegA, c</em>, where c is some integer - then the usual assembler instructions looks like this:</p>
<div class="asm" style="font-family:monospace;color: #000066;  border: solid thin #c2c1b1; background: #d6d5c5;">&nbsp; <span style="color: #00007f; font-weight: bold;">mov</span> Reg2<span style="color: #339933;">,</span> RegA<br/>
&nbsp; <span style="color: #00007f; font-weight: bold;">shl</span> Reg2<span style="color: #339933;">,</span> c<br/>
&nbsp; <span style="color: #00007f; font-weight: bold;">shr</span> RegA<span style="color: #339933;">,</span> <span style="color: black;">&#40;</span><span style="color: #ff0000;">32</span> <span style="color: #339933;">-</span> c<span style="color: black;">&#41;</span><br/>
&nbsp; <span style="color: #00007f; font-weight: bold;">xor</span> RegA<span style="color: #339933;">,</span> Reg2<br/>
&nbsp;</div>
<p>Translating this into the ArtEvol MetaLanguage is very easy:</p>
<div class="asm" style="font-family:monospace;color: #000066;  border: solid thin #c2c1b1; background: #d6d5c5;">&nbsp; zer0<br/>
&nbsp; addnumber c &nbsp; <span style="color: black; font-style: italic;">; BC1=c</span><br/>
&nbsp; save &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; BC2=c</span><br/>
<br/>
&nbsp; nopsA &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; BC1=RegA</span><br/>
&nbsp; <span style="color: #00007f; font-weight: bold;">shl</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; BC1=BC1 shl BC2 = RegA shl c</span><br/>
<br/>
&nbsp; <span style="color: #00007f; font-weight: bold;">push</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; temporarly save at stack</span><br/>
<br/>
&nbsp; zer0<br/>
&nbsp; addnumber <span style="color: black;">&#40;</span><span style="color: #ff0000;">32</span><span style="color: #339933;">-</span>c<span style="color: black;">&#41;</span> <span style="color: black; font-style: italic;">; BC1=(32-c)</span><br/>
&nbsp; save &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; BC2=(32-c)</span><br/>
<br/>
&nbsp; nopsA &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; BC1=RegA</span><br/>
&nbsp; <span style="color: #00007f; font-weight: bold;">shr</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; BC1=BC1 shr BC2 = RegA shr (32-c)</span><br/>
<br/>
&nbsp; save &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; BC2=RegA shr (32-c)</span><br/>
<br/>
&nbsp; <span style="color: #00007f; font-weight: bold;">pop</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; restore value (RegA shl c)</span><br/>
&nbsp; <span style="color: #00007f; font-weight: bold;">xor</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; BC1 xor BC2 = (RegA shl c) XOR (RegA shr (32-c))</span><br/>
<br/>
&nbsp; nopdA &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; RegA=(RegA shl c) XOR (RegA shr (32-c))</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; = RegA rol c &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </span><br/>
&nbsp;</div>
<p>The 2nd argument (c in this case) is saved in the BC2 register, then the first argument is loaded into the BC1 register and the operation is performed. As ROL requires three operations, the result of the first one is temporarily saved at the stack; the same procedure is performed again, and in the end, the results are combined.</p>
<p>The <em>addnumber</em> is a macro which returns the right combination of <em>addNNNN</em> instructions.</p>
<h3><a name="c22"></a>2.2 Translator</h3>
<p>In order to convert the metalanguage instructions to native x86 instructions, a tiny translator is used:</p>
<div class="asm" style="font-family:monospace;color: #000066;  border: solid thin #c2c1b1; background: #d6d5c5;">&nbsp;<span style="color: #339933;">.</span><span style="color: #0000ff; font-weight: bold;">code</span><br/>
&nbsp;<span style="color: #0000ff; font-weight: bold;">start</span><span style="color: #339933;">:</span><br/>
&nbsp; &nbsp; &nbsp; invoke &nbsp;VirtualAlloc<span style="color: #339933;">,</span> <span style="color: #ff0000;">0x0</span><span style="color: #339933;">,</span> <span style="color: black;">&#40;</span>EndAmino<span style="color: #339933;">-</span>StAmino<span style="color: black;">&#41;</span><span style="color: #339933;">*</span><span style="color: #ff0000;">8</span><span style="color: #339933;">,</span> \<br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #ff0000;">0x1000</span><span style="color: #339933;">,</span> PAGE_EXECUTE_READWRITE<br/>
<br/>
&nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">mov</span> &nbsp; &nbsp; <span style="color: black;">&#91;</span>Place4Life<span style="color: black;">&#93;</span><span style="color: #339933;">,</span> <span style="color: #46aa03; font-weight: bold;">eax</span><br/>
<br/>
&nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">mov</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">edx</span><span style="color: #339933;">,</span> <span style="color: #ff0000;">0x0</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; EDX ... codon counter</span><br/>
&nbsp; &nbsp; &nbsp; WriteMoreToMemory<span style="color: #339933;">:</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">mov</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">ebx</span><span style="color: #339933;">,</span> <span style="color: #ff0000;">0x0</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; EBX = 0</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">mov</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">bl</span><span style="color: #339933;">,</span> <span style="color: #0000ff; font-weight: bold;">byte</span><span style="color: black;">&#91;</span><span style="color: #46aa03; font-weight: bold;">edx</span><span style="color: #339933;">+</span>StAmino<span style="color: black;">&#93;</span> <span style="color: black; font-style: italic;">; BL &nbsp;= current codon</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">shl</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">ebx</span><span style="color: #339933;">,</span> <span style="color: #ff0000;">3</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; EBX*= 8</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">mov</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">esi</span><span style="color: #339933;">,</span> StartAlphabet &nbsp; &nbsp;<span style="color: black; font-style: italic;">; ESI = alphabet offset</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">add</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">esi</span><span style="color: #339933;">,</span> <span style="color: #46aa03; font-weight: bold;">ebx</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; ESI = offset for current</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; &nbsp; &nbsp; &nbsp; amino acid</span><br/>
<br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">mov</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">ebx</span><span style="color: #339933;">,</span> <span style="color: #46aa03; font-weight: bold;">edx</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; EBX = codon counter</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">shl</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">ebx</span><span style="color: #339933;">,</span> <span style="color: #ff0000;">3</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; EBX*=8 = size(amino acids)</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">mov</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">edi</span><span style="color: #339933;">,</span> <span style="color: black;">&#91;</span>Place4Life<span style="color: black;">&#93;</span> &nbsp; &nbsp; <span style="color: black; font-style: italic;">; EDI = memory address</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">add</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">edi</span><span style="color: #339933;">,</span> <span style="color: #46aa03; font-weight: bold;">ebx</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; memory offset of current</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; amino acid</span><br/>
<br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">mov</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">ecx</span><span style="color: #339933;">,</span> <span style="color: #ff0000;">8</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; ECX=8</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">rep</span> &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">movsb</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; write 8 byte from alphabet</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; to memory</span><br/>
<br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">inc</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">edx</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; increase counter</span><br/>
<br/>
&nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">cmp</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">edx</span><span style="color: #339933;">,</span> <span style="color: black;">&#40;</span>EndAmino<span style="color: #339933;">-</span>StAmino<span style="color: black;">&#41;</span><br/>
&nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">jne</span> &nbsp; &nbsp; WriteMoreToMemory &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; transformed all codons?</span><br/>
<br/>
&nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">call</span> &nbsp; &nbsp;<span style="color: black;">&#91;</span>Place4Life<span style="color: black;">&#93;</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; let's start!!!</span><br/>
<br/>
<br/>
&nbsp;StartAlphabet<span style="color: #339933;">:</span><br/>
&nbsp; &nbsp; &nbsp; _pushall <span style="color: #0000ff; font-weight: bold;">EQU</span> <span style="color: #ff0000;">0</span> &nbsp; &nbsp;<span style="color: black; font-style: italic;">; 0000 0000 - 0:</span><br/>
&nbsp; &nbsp; &nbsp; Command0<span style="color: #339933;">:</span> &nbsp;<span style="color: #00007f; font-weight: bold;">pushad</span><br/>
&nbsp; &nbsp; &nbsp; <span style="color: #0000ff; font-weight: bold;">times</span> <span style="color: black;">&#40;</span><span style="color: #ff0000;">8</span><span style="color: #339933;">-</span><span style="color: #0000ff; font-weight: bold;">$</span><span style="color: #339933;">+</span>Command0<span style="color: black;">&#41;</span><span style="color: #339933;">:</span> <span style="color: #00007f; font-weight: bold;">nop</span><br/>
<br/>
&nbsp; &nbsp; &nbsp; _nopsA <span style="color: #0000ff; font-weight: bold;">EQU</span> <span style="color: #ff0000;">1</span> &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; 0000 0001 - 1:</span><br/>
&nbsp; &nbsp; &nbsp; Command1<span style="color: #339933;">:</span> &nbsp;<span style="color: #00007f; font-weight: bold;">mov</span> <span style="color: #46aa03; font-weight: bold;">ebx</span><span style="color: #339933;">,</span> <span style="color: #46aa03; font-weight: bold;">eax</span><br/>
&nbsp; &nbsp; &nbsp; <span style="color: #0000ff; font-weight: bold;">times</span> <span style="color: black;">&#40;</span><span style="color: #ff0000;">8</span><span style="color: #339933;">-</span><span style="color: #0000ff; font-weight: bold;">$</span><span style="color: #339933;">+</span>Command1<span style="color: black;">&#41;</span><span style="color: #339933;">:</span> <span style="color: #00007f; font-weight: bold;">nop</span><br/>
<br/>
&nbsp;<span style="color: black; font-style: italic;">;[...]</span><br/>
<br/>
&nbsp;StAmino<span style="color: #339933;">:</span><br/>
&nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; Code written in the ArtEvol metalanguage</span><br/>
<br/>
&nbsp;EndAmino<span style="color: #339933;">:</span><br/>
&nbsp;<span style="color: #339933;">.</span>end <span style="color: #0000ff; font-weight: bold;">start</span><br/>
&nbsp;</div>
<h3><a name="c23"></a>2.3 Replication, Mutation and Selection</h3>
<p>To achieve Evolution, a system requires Reproduction, Mutations and Selection.</p>
<p>In the <strong>reproduction</strong> stage, the organism creates a living copy of itself. In the artificial system of a computer, this can be done in the same way as computer viruses and computer worms do - interfere with special file formats or network protocols such that the copy will be executed in a different habitat (other computer, other file, ...). However, this requires alot of previous knowledge about the system, thus is not the simplest starting point for evolution. The most trivial way of reproduction is to copy the own file in the current directory and run it - which is actually the way how it is done in the experiments.</p>
<p>The interference between reproduction and <strong>mutations</strong> leads to non-identical replica of the organism. In the biological system, mutations happens due to disruptive effect such as cosmic X-rays. This leads to point mutations (exchange of one single codon) or more difficult chromosome abnormality. The natural mutation probability in a computer system (such as mistakes in the copy process) is neglectable, thus the organism has to carry its own mutation engines. This is explained in detail in the next chapter.</p>
<p>Natural <strong>selection</strong> is a process in which a certain trait becomes more or less common in the population, depending on its effect on the fitness of the organism. This process appears when the organism compete, struggle for limited values (such as energy), or are exposed to natural enemies. In the artificial system of a computer, anti virus programs could be responsible for natural selection (and by that unwillingly initiate a faster evolutionary process at all). A different selective pressure comes from attentive user, who would stop any suspicious behaviour.</p>
<h2><a name="c3"></a>3 Mutations</h2>
<h3><a name="c31"></a>3.1 Point mutation - bit flip</h3>
<p>Point mutations change single nucleobases in the DNA. These mutations can be categorized into silent (when the affected codon maps to the same amino acid due to the redundance in the alphabet), missense (when the affected codon maps to a different amino acid); or nonsense (when the affected codon maps to the STOP codon).</p>
<p>A native analogon to that concept would be the change of single bits in the organism - called <em>Bitflip</em>. These mutations have the same categories as their biological companion.</p>
<p>The mutation rate (mutations per base per generation) in biological organisms varies from 10^-4 for very small (some kilo bases) to 10^-8 - 10^-9 for humans (some giga bases). Finding an adequate mutation rate for artificial organisms is not trivial, as too small values lead to mainly unmutated offspring, thus no evolution; whereas too big mutation rates lead to extinction of the population. To get the best, one could test organisms with different mutation rate, and take the critical value.</p>
<div align="center">
<img src="img/vsp37/MutaRate07001.PNG" alt="Figure 3.1: Five populations with a mutation rate of 1 / 7.001"/>
<p><strong>Figure 3.1: Five populations with a mutation rate of 1 / 7.001</strong></p>
<img src="img/vsp37/MutaRate09001.PNG" alt="Figure 3.2: Five populations with a mutation rate of 1 / 9.001"/>
<p><strong>Figure 3.2: Five populations with a mutation rate of 1 / 9.001</strong></p>
<img src="img/vsp37/MutaRate11003.PNG" alt="Figure 3.3: Five populations with a mutation rate of 1 / 11.003"/>
<p><strong>Figure 3.3: Five populations with a mutation rate of 1 / 11.003</strong></p>
<img src="img/vsp37/MutaRate13499.PNG" alt="Figure 3.4: Five populations with a mutation rate of 1 / 13.499"/>
<p><strong>Figure 3.4: Five populations with a mutation rate of 1 / 13.499</strong></p>
</div>
<p>The test-organism - which can create three offspring - has a size of 20.480 bytes, the critical mutation rate is between 1 / 9.001 and 1 / 11.003 ; the probability that at least one bit is changed is between 84.5% and 89.7%.</p>
<h3><a name="c32"></a>3.2 Chromosomal inversion - byte eXCHanGe</h3>
<p>A different kind of mutation happens when a segment of a chromosome is reversed. This happens when a segment breaks off and is rearranged in the wrong way.</p>
<p>A similar method is used in this project: two consecutive d-words (that means, 4 codons each) are exchanged. This mutation is not as dangerous as it may look like at first; the codon streams forming a functional part are very big, thus are not too sensitive on such local translocations.</p>
<div align="center">
<img src="img/vsp37/MutaRate26501_xchg.PNG" alt="Figure 3.5: Five populations with a mutation rate of 1 / 25.501"/>
<p><strong>Figure 3.5: Five populations with a mutation rate of 1 / 25.501</strong></p>
<img src="img/vsp37/MutaRate27011_xchg.PNG" alt="Figure 3.6: Five populations with a mutation rate of 1 / 27.011"/>
<p><strong>Figure 3.6: Five populations with a mutation rate of 1 / 27.011</strong></p>
<img src="img/vsp37/MutaRate27253_xchg.PNG" alt="Figure 3.7: Five populations with a mutation rate of 1 / 27.253"/>
<p><strong>Figure 3.7: Five populations with a mutation rate of 1 / 27.253</strong></p>
<img src="img/vsp37/MutaRate27509_xchg.PNG" alt="Figure 3.8: Five populations with a mutation rate of 1 / 27.509"/>
<p><strong>Figure 3.8: Five populations with a mutation rate of 1 / 27.509</strong></p>
</div>
<p>These graphs show that the critical mutation rate is much sharper than for Bit flips - this is what you would expect as Byte Exchange has much stronger effects.</p>
<h3><a name="c33"></a>3.3 Deletion, insertion, translocation</h3>
<p>Deletion is a mutation in which a part of the DNA is missing; insertion is the inverse process, where an additional sequence of DNA is included into the genom. Translocation is a combination of these mutations: A part of the DNA breaks off and is included at a different place.</p>
<p>A similar method can be realized in artificial organisms in the computer system within one simple algorithm. Three random values are calculated (the place of the mutation P, the size of the inserted NOP block S<sub>i</sub> and the size of the translocated block S<sub>b</sub>). Then at P, a block with the size of S<sub>b</sub> will be translocated S<sub>i</sub> bytes, the new created block at P will be filled with NOPs.</p>
<div align="center">
<img src="img/vsp37/deletion.png" width="540" height="133" alt="Figure 3.9: Deletion process in the organism."/>
<p><strong>Figure 3.9: Deletion process in the organism. Grey are NOPs, red and orange are functional parts. Red is not moved, orange is the translocated sequence. The second red part is smaller after deletion, as it has lost some of its code.</strong></p>
<img src="img/vsp37/insertion.png" width="540" height="133" alt="Figure 3.10: Insertion process in the organism."/>
<p><strong>Figure 3.10:Insertion process in the organism. Grey are NOPs, red and orange are functional parts. Red is not moved, orange is the translocated sequence. The whole orange part is translocated, thus there is nothing deleted - just an insertion of NOPs.</strong></p>
</div>
<h3><a name="c34"></a>3.4 Horizontal gene transfer</h3>
<p>Horizontal gene transfer is a process in which an organism incorporates genetic material from another organism without being the offspring of that organism. In biological systems, this is a controlled (not by random mutations) method to receive beneficial functions such as antibiotic resistance. Photosynthesis is an important process which has been developed with horizontal gene transfer from different bacteria.</p>
<p>In the artificial system, an organism could try to interact with other organism and exchange valid code, and therefore perform a symbiotic conjugation. However, this would require a specific protocol for communication (such as F-plasmids in bacteria) - which is not developed so far.</p>
<p>Nevertheless the artificial organism could try to gain new information from other files, just by opening them and copy some parts of their code. In the case that the other file is written in the same language, the organism has the chance of getting new functions.</p>
<h3><a name="c35"></a>3.5 Polymorphism: neutral codon variation</h3>
<p>In the artificial organisms, the alphabet has 2^8 = 256 entries which map to 45 or less instructions, thus there is a big redundancy - that means several codons map to the same amino acid.</p>
<p>The organism can scan thru its alphabet, detect equal amino acids, then scan its codon-stream and exchange the codons which point to the same amino acid.</p>
<p>There are a few advantages to use this technique in artificial organism: Firstly, codons which point to isolated amino acids (these who can not be transformed by a single bit flip to another amino acid of the same kind) can be de-isolated, thereby increase the robustness of the overall code. Secondly, such macro mutations are of high importance to bypass natural enemies (such as antivirus software), thus increase fitness. And thirdly, mutations in the polymorphism-engine itself or variations of the START or STOP codon could lead to unpredictable results.</p>
<h2><a name="c4"></a>4 Further improvements</h2>
<h3><a name="c41"></a>4.1 Start- and Stop codons: Splicing</h3>
<p>Natural genetic code contains alot of non-functional garbage, which can be old unused DNA or (malformed) duplicates of actual functional code. In human DNA, approximately 95% of the DNA is garbage. These non-coding parts are called Introns, the functional parts are Exons. Before translation of mRNA into proteins, the introns are cutted out in a process called Splicing - this is done by taking usage of two special codons - the START and the STOP codon. Each functional part starts with a START-codon and ends with a STOP-codon - the parts between a STOP codon and the next START codon (which is actually the intron) will be removed.</p>
<p>The advantage of such introns is that within the unused code new DNA sequences could be developed which are (in very rare cases) actually functional - and have different behaviour. Together with the additional possibility for altering its code, this would be a special gain for artificial organisms.</p>
<div align="center">
<img src="img/vsp37/splicing.png" width="835" height="210" alt="Figure 4.1: Splicing functionality in a artificial organism"/>
<p><strong>Figure 4.1: Splicing functionality in a artificial organism</strong></p>
</div>
<p>Implementing a very small splicing algorithm into the translator can be achieved in the following way:</p>
<ul>
<li>All codons in the form of <strong>1??1.???1</strong> (32 codons) point to a NOP amino acid</li>
<li>At the translation, whenever there is a STOP codon, AL=0x91 (1001.0001); whenever there is a START codon, AL=0x0</li>
<li>Each codon will be OR'ed with AL</li>
<li>In the end, each codon after a STOP mark will be redirected to a NOP amino acid - until there is a START codon</li>
</ul>
<div class="asm" style="font-family:monospace;color: #000066;  border: solid thin #c2c1b1; background: #d6d5c5;">&nbsp; &nbsp; &nbsp; invoke &nbsp;VirtualAlloc<span style="color: #339933;">,</span> <span style="color: #ff0000;">0x0</span><span style="color: #339933;">,</span> AlignedSize<span style="color: #339933;">,</span> \<br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #ff0000;">0x1000</span><span style="color: #339933;">,</span> PAGE_EXECUTE_READWRITE<br/>
&nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">mov</span> &nbsp; &nbsp; <span style="color: black;">&#91;</span>Place4Life<span style="color: black;">&#93;</span><span style="color: #339933;">,</span> <span style="color: #46aa03; font-weight: bold;">eax</span><br/>
<br/>
&nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">mov</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">eax</span><span style="color: #339933;">,</span> <span style="color: #ff0000;">0x0</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; EAX ... Splicing seperator</span><br/>
&nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">mov</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">edx</span><span style="color: #339933;">,</span> <span style="color: #ff0000;">0x0</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; EDX ... codon counter</span><br/>
&nbsp; &nbsp; &nbsp; WriteMoreToMemory<span style="color: #339933;">:</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">mov</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">ebx</span><span style="color: #339933;">,</span> <span style="color: #ff0000;">0x0</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">mov</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">bl</span><span style="color: #339933;">,</span> <span style="color: #0000ff; font-weight: bold;">byte</span><span style="color: black;">&#91;</span><span style="color: #46aa03; font-weight: bold;">edx</span><span style="color: #339933;">+</span>StAmino<span style="color: black;">&#93;</span> <span style="color: black; font-style: italic;">; BL = current codon</span><br/>
<br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">cmp</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">bl</span><span style="color: #339933;">,</span> StartCodon &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; is it the Start Codon?</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">jne</span> &nbsp; &nbsp; SplicingNoStart<br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">mov</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">eax</span><span style="color: #339933;">,</span> <span style="color: #ff0000;">0x0</span> &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; yes -&gt; eax=0x0</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; SplicingNoStart<span style="color: #339933;">:</span><br/>
<br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">cmp</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">bl</span><span style="color: #339933;">,</span> StopCodon &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; is it the Stop codon?</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">jne</span> &nbsp; &nbsp; SplicingNoStop<br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">mov</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">eax</span><span style="color: #339933;">,</span> <span style="color: #ff0000;">0x91</span> &nbsp; &nbsp; <span style="color: black; font-style: italic;">; yes -&gt; eax = 0x91</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; SplicingNoStop<span style="color: #339933;">:</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;= 1001 0001</span><br/>
<br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">or</span> &nbsp; &nbsp; &nbsp;<span style="color: #46aa03; font-weight: bold;">bl</span><span style="color: #339933;">,</span> <span style="color: #46aa03; font-weight: bold;">al</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; SPLICING!</span><br/>
<br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">shl</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">ebx</span><span style="color: #339933;">,</span> <span style="color: #ff0000;">3</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; EBX*=8;</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">mov</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">esi</span><span style="color: #339933;">,</span> StartAlphabeth &nbsp; <span style="color: black; font-style: italic;">; ESI = alphabet offset</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">add</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">esi</span><span style="color: #339933;">,</span> <span style="color: #46aa03; font-weight: bold;">ebx</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; ESI = offset for current</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; &nbsp; &nbsp; &nbsp; amino acid</span><br/>
<br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">mov</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">ebx</span><span style="color: #339933;">,</span> <span style="color: #46aa03; font-weight: bold;">edx</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; EBX = codon counter</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">shl</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">ebx</span><span style="color: #339933;">,</span> <span style="color: #ff0000;">3</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; EBX*=8 = size(amino acids)</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">mov</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">edi</span><span style="color: #339933;">,</span> <span style="color: black;">&#91;</span>Place4Life<span style="color: black;">&#93;</span> &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; EDI = memory address</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">add</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">edi</span><span style="color: #339933;">,</span> <span style="color: #46aa03; font-weight: bold;">ebx</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; memory offset of current</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; amino acid</span><br/>
<br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">mov</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">ecx</span><span style="color: #339933;">,</span> <span style="color: #ff0000;">8</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; ECX=8</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">rep</span> &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">movsb</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; write 8 bytes from alphabet</span><br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; to memory</span><br/>
<br/>
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">inc</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">edx</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; increase counter</span><br/>
<br/>
&nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">cmp</span> &nbsp; &nbsp; <span style="color: #46aa03; font-weight: bold;">edx</span><span style="color: #339933;">,</span> <span style="color: black;">&#40;</span>EndAmino<span style="color: #339933;">-</span>StAmino<span style="color: black;">&#41;</span><br/>
&nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">jne</span> &nbsp; &nbsp; WriteMoreToMemory &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <span style="color: black; font-style: italic;">; transformed all codons?</span><br/>
<br/>
&nbsp; &nbsp; &nbsp; <span style="color: #00007f; font-weight: bold;">call</span> &nbsp; &nbsp;<span style="color: black;">&#91;</span>Place4Life<span style="color: black;">&#93;</span> &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;<span style="color: black; font-style: italic;">; let's start!!!</span><br/>
&nbsp;</div>
<h3><a name="c42"></a>4.2 Optimization of the instruction set</h3>
<p>The original instruction set had 43 different entries. Ofria, Adami and Collier found out, that a smaller instruction set leads to higher robustness under mutations, thus higher fitness[[<a href="#f6_6" name="b6_6">6</a>]].</p>
<p>In their experiments, the reason is that small instruction set requires bigger realisations of function, thus the risk of a lethal mutation is spread over a larger area. In our realisation, a second advantage appeares. A small instruction set leads to a more redundant alphabet, therefore allows more codons to point to the same amino acid. In the end, there is a bigger probability that a single BitFlip changes the codon such that it still points to the original amino acid.</p>
<p>Peter Ferrie was able to create an optimized instruction set with just 18 entries, by replacing an instruction with a combination of other instructions.[<a href="#f11_11">11</a>] One simple example is <em>zer0 -> save+xor</em>.</p>
<p>For implementing these optimizations, one has to take care of changed Registers and Flags. This can be done by using Stack instruction (<em>pushall</em> and <em>popall</em>) or even implement a new <em>pseudo</em>register (in the .data section) with some special properties.</p>
<p>Unfortunately, these implementations lead to an excessive useage of evolutionary <em>dangerous</em> instructions, which are instructions that lead to the programs crash if they are replaced by other instructions.</p>
<p>I consider dangerous instructions as everything that interacts with the stack (<em>push, pop, pushall, popall, CallAPILoadLibrary</em>), that in fluences the code flow (<em>JnzDown, JnzUp, call, saveJmpOff</em>), and that interacts with the memory (saveWrtOff, writeByte, writeDWord, getdata). One can create two further categories: semi-harmless, which are all instructions that change the values of <em>RegA, RegB, RegD and BC2</em>, and harmless instructions just change the BC1 register.</p>
<div align="center">
<img src="img/vsp37/OptInstrALL.PNG" alt="Figure 4.2: Original instruction set"/>
<p><strong>Figure 4.2: Original instruction set</strong></p>
<img src="img/vsp37/OptInstrzer0.PNG" alt="Figure 4.3: zer0 removed"/>
<p><strong>Figure 4.3: zer0 removed</strong></p>
<img src="img/vsp37/OptInstrsubsaved.PNG" alt="Figure 4.4: subsaved removed"/>
<p><strong>Figure 4.4: subsaved removed</strong></p>
<img src="img/vsp37/OptInstraddNNNN.PNG" alt="Figure 4.5: addNNNN removed"/>
<p><strong>Figure 4.5: addNNNN removed</strong></p>
<img src="img/vsp37/OptInstraddsaved.PNG" alt="Figure 4.6: addsaved removed"/>
<p><strong>Figure 4.6: addsaved removed</strong></p>
<img src="img/vsp37/OptInstradd0001.PNG" alt="Figure 4.7: add0001 removed"/>
<p><strong>Figure 4.7: add0001 removed</strong></p>
</div>
<p>The x-axis shows the different instructions (with the same order as written in chapter 2), the y-axis gives the (normalized) appearence of the instruction in an organism. It is interesting to see how dangerous instruction density increases, especially after removing the addsaved instruction.</p>
<h3><a name="c43"></a>4.3 Optimization of the alphabet</h3>
<p>To achieve the optimal robustness, evolution has lead to a special order of codon mapping. As an example, the amino acid Proline can be coded via CCU, CCC, CCA and CCG. That means, whenever a mutation changes the third nucleobase, still a codon remains that codes Prolin (in biological systems, mutations happens most often at the third nucleobase).</p>
<p>For artificial organisms, one has 256 slots for about 45 instructions, and furthermore different pairs of exchanged codons have different probability to cause mistakes (exchanging <em>add0001</em> with <em>add0004</em> may cause less problems than exchanging <em>add0001</em> with <em>pushall</em>).</p>
<p>This is a nonlinear problem, and solutions by hand take long and are of low quality. However, one can reformulate the problem - with the help of a bit of physics:</p>
<p>Let's imagine the codons as objects in a special space, such that each two codons with one bit difference are neighbors (the geometry of this space is an 8 dimensional cube with codons on the corners). For example, the codon 0000.0000 and 0010.0000 are neighbors. Each codon interacts with its neighbors, thus has an interaction energy (which depends on the types of the codons).</p>
<p>We can define V(A,B) as the interaction energy of codon A and B. One possible definition would be</p>
<table summary="Interaction Energy">
<tr><td>V(A,B) = 0</td><td>if A = B</td></tr>
<tr><td>V(A,B) = 0.5</td><td>if A and B are addNNNN or sub0001</td></tr>
<tr><td>V(A,B) = 0.66</td><td>if A and B are harmless instructions</td></tr>
<tr><td>V(A,B) = 0.75</td><td>if A and B are harmless or semi-harmless instructions</td></tr>
<tr><td>V(A,B) = 1</td><td>else (if A or B is a dangerous instruction or a START or STOP codon)</td></tr>
</table>
<p>Now we can define the total energy of the system as:</p>
<p><img src="/img/cache/b07d1b12f0bb691eaffbe0ace7e01bde.gif" alt="
\begin{align}
E_{\text{total}}=\sum_{i=1}^{256} \sum_{j=1}^8 V(\text{codon}_i, \text{codon}_j)\nonumber
\end{align}
" valign="middle"/></p>
<p>The total energy of a maximum random system would be 2.048, the mimumum total energy of a system with just one single instruction would be zero.</p>
<p>Finally, we can reformulate "find an optimal alphabet" into "Find the minimal interaction energy of the system". There are several ways to find a minimum energy, one is the Metropolis-Algorithm; we use a slightly modified one.</p>
<p>First, we fill the 256 entries with instructions of random order, and calculate the total energy of the system. Then we exchange a few instructions randomly, and calculate the new total energy. If the new energy is smaller than the old one, we keep the new system, otherwise we continue with the old one. To find a (local) minimum, one can repeat that method.</p>
<div align="center">
<img src="img/vsp37/energy.PNG" alt="Figure 4.8: Energy minimization of five systems"/>
<p><strong>Figure 4.8: Energy minimization of five systems. x-axis is iteration number (in 1000s), y-axis is energy.</strong></p>
</div>
<p>The figure shows that the alphabet reaches a good local minimum after a few 100.000 iterations. It also shows that systems with different starting order find local minimal of approximately the same energy.</p>
<h2><a name="c5"></a>5 Experiments</h2>
<p>An experiment measures the fitness of the organisms by letting them struggle for limited resources (such as CPU time and memory). To control the experiment, several guard files ran in the background, to close endless-loop files, multiple instances of the same file, unmutated files by a certain probability. A more detailed explanation can be found in [<a href="#f8_8">8</a>].</p>
<h3><a name="c51"></a>5.1 Hamming distance</h3>
<p>In the first experiment, we analyse the <em>long time</em> behaviour of a population, that can just perform BitFlips and Byte XCHG.</p>
<p>The Hamming distance (difference in the bit-code) with respect to the original ancestor is calculated every 3 minutes.</p>
<div align="center">
<img src="img/vsp37/hamming.PNG" alt="Figure 5.1: Evolution of Hamming distance over 12 hours."/>
<p><strong>Figure 5.1: Evolution of Hamming distance over 12 hours. x-axis: Hamming distance, y-axis: time (in minutes)</strong></p>
</div>
<p>It is very surprising and interesting, that after about 7.5 hours a mutation has occured with the effect that some organisms can largely bypass the no-cloning guard. A deeper analysis of that event would require reverse engineering of mutated code, which is a non-trivial task - and therefore hasn't been done yet.</p>
<p>However, this event is an indication that artificial organisms can bypass control instances very fast.</p>
<p>A second experiment has been performed, with an organism which contains about 90% introns (similar to natural organisms).</p>
<div align="center">
<img src="img/vsp37/hamming_intron.PNG" alt="Figure 5.2: Evolution of Hamming distance for an organism with a high amount of introns over 6 hours."/>
<p><strong>Figure 5.2: Evolution of Hamming distance for an organism with a high amount of introns over 6 hours. x-axis: Hamming distance, y-axis: time (in minutes)</strong></p>
</div>
<p>As one would expect, the mutation rate is much higher and very constant, the standard derivation of the hamming distance spreads continuously.</p>
<h3><a name="c52"></a>5.2 Effect of alphabet "energy"</h3>
<p>As explained in chapter 4.3, each alphabet has a specific "energy". To see whether the used defnition of the interaction actually give a good result for robust alphabets, an evolutionary experiment has been performed.</p>
<div align="center">
<img src="img/vsp37/AlphabetLvsRnd1.PNG" alt="Figure 5.3: Energy 1662.9"/>
<p><strong>Figure 5.3: Energy 1152.1 (blue) vs. 1662.9 +/- 11.2 (red)</strong></p>
<img src="img/vsp37/AlphabetLvsRnd2.PNG" alt="Figure 5.4: 1386.8"/>
<p><strong>Figure 5.4: Energy 1152.1 (blue) vs. 1386.8 +/- 3.2 (red)</strong></p>
</div>
<p>Figure 5.3 and 5.4 show that lower energy has a better fitness, thus the presented interaction definition is at least roughly a good approximation for an optimal alphabet.</p>
<h3><a name="c53"></a>5.3 Effect of Start- and Stop-codons</h3>
<p>One experiment has been performed with a number of 100.000 random codons within a STOP and a START codon - a big intron. Several organisms have converted a part of the intron into an exon (by introducing a START codon) without any negative effect - all converted functional parts were neutral mutations.</p>
<p>There were two very surprising results: The biggest converted part had 33 codons, and still worked without problem. A different converted exon even managed to perform an API call without crashing.</p>
<div align="center">
<img src="img/vsp37/intronToexon.PNG" alt="Introns to Exons"/>
</div>
<p>Such huge blocks of non-lethal codons and API calls could be very valuable for artificial organisms, which are chased by behaviour scanners and API call tracers.</p>
<h2><a name="c6"></a>6 Outlook</h2>
<p>There are several features in the natural protein biosynthesis (or in microbiology in general) which could be used in this artificial concept:</p>
<p><strong>Alternative Splicing</strong>: Before translating the codons into amino acids, the splicing process cuts out introns. In natural systems, there can be an alternative splicing process, which can create the final codon sequence in many alternative ways. For example, exons can be combined in different ways, some exons could be cutted out, introns can be coded, and other methods. This process is in uenced by regulatory elements (such as proteins). An analogy to this process would increase the variability of the organisms drastically.</p>
<p><strong>Protein Folding</strong>: After translation of codons into amino acids, a process called folding gives a specific 3D structure to the amino acid chain - this structure is mainly responsible for the proteins chemical properties. A similar further layer of translation may have advantages for the organisms too.</p>
<p><strong>Protocol for Horizontal Gene Transfer</strong>: To exchange valuable information as antibiotic resistance, bacteria have developed processes called conjugation. To develope a process like that for exchanging useful information within the population would be a big advantage.</p>
<p>There are some further techniques which would increase the variability of the organisms, such as increasing the functional code size. The diffculties come from restrictions of the PE format, which requires more than one field to be mutated in the same manner. This is a very unlikely process - finding a solution to that problem would open many new possibilities for the organisms.</p>
<p>The experiments have shown a very promising behaviour of the different mutation techniques. Especially the START- and STOP-codon experiment, where organisms performed an additional API call, indicates that even macro mutations are realistic within this concept - and behaviour changing mutations are not always lethal.</p>
<p>The question of how antivirus programs can detect organisms using this technique is open. The organism's non-lethal configuration is infinite, and due to the fact that darwinian evolution is not predictable, algorithmic approaches are probably unusable and limited. Behaviour scanners are likely (due to effects that has been shown with Splicing) not practicable, too. Statistical approaches may work for a low number of generations quite good, but will most likely fail for big difference to the ancestor (at high generations or when several macro mutations happened), as well.[<a href="#f12_12" name="b12_12">12</a>]</p>
<p>... as a conclusion, one can see:</p>
<p><strong>The artificial organisms took the redpill - and enjoy their new freedom now...</strong></p>
<h2><a name="c7"></a>7 References</h2>
<p><a name="f1_1" href="#b1_1">1.</a> Thomas S. Ray, <em>An approach to the synthesis of life</em>, Artificial Life II, vol. XI of SFI Studies in the Sciences of Complexity, p.371-408, 1992.</p>
<p><a name="f2_2" href="#b2_2">2.</a> Kurt Thearling and Thomas S. Ray, <em>Evolving Multi-cellular Artificial Life</em>, Proceedings of Artificial Life IV, MIT Press: Cambridge, 1994.</p>
<p><a name="f3_3" href="#b3_3">3.</a> Kurt Thearling and Thomas S. Ray, <em>Evolving Parallel Computation</em>, Complex Systems, Volume 10, Number 3,1997.</p>
<p><a name="f4_4" href="#b4_4">4.</a> Claus O. Wilke, Jia Lan Wang, Charles Ofria, Richard E. Lenski and Christoph Adami, <em>Evolution of digital organisms at high mutation rates leads to survival of the flattest</em>, Nature, 412:331-333, 2001.</p>
<p><a name="f5_5" href="#b5_5">5.</a> Richard E. Lenski, Charles Ofria, Robert T. Pennock and Christoph Adami, <em>The Evolutionary Origin of Complex Features</em>, Nature, 423:139-144, 2003.</p>
<p><a name="f6_6" href="#b6_6">6.</a> Charles Ofria, Christoph Adami, and Travis C. Collier, <em>Design of Evolvable Computer Languages</em>, IEEE Transactions on Evolutionary Computation, vol. 6, no. 4, p.420-424, 2002.</p>
<p><a name="f7_7" href="#b7_7">7.</a> Dimitris Iliopoulos, Christoph Adami and Peter Ször, <em><a href="/lib/aps01.html">Darwin inside the machines: Malware evolution and the consequences for computer security</a></em>, Proceedings of Virus Bulletin Conference 2008, p.187-194, 2008.</p>
<p><a name="f8_8" href="#b8_8">8.</a> Thomas Sperl, <em><a href="/lib/vsp26.html">Taking the redpill: Artificial Evolution in native x86 systems</a></em>, October 2010.</p>
<p><a name="f9_9" href="#b9_9">9.</a> Thomas Sperl, <em>Mutational Robustness in x86 systems</em>, November 2010.</p>
<p><a name="f10_10" href="#b10_10">10.</a> Peter Ferrie, <em><a href="/lib/apf56.html">Flibi Night</a></em>, Virus Bulletin March 2011, p.4-5, 2011.</p>
<p><a name="f11_11" href="#b11_11">11.</a> Peter Ferrie, <em><a href="/lib/apf57.html">Flibi: Evolution</a></em>, Virus Bulletin May 2011, p.6-15, 2011.</p>
<p><a name="f12_12" href="#b12_12">12.</a> Private communication with Mostafa Saleh in June 2011.</p>
[<a style="" href="/lib/?lang=EN&amp;index=AI#vsp37">Back to index</a>] [<a href="/lib/vsp37.html#disqus_thread">Comments</a>]<br/> <div id="disqus_thread"></div>
<script type="text/rocketscript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'vxheaven'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
<noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript></div>
<div><small>By accessing, viewing, downloading or otherwise using this content you agree to be bound by the <a href="/agreement.php">Terms of Use</a>!</small> <small>vxheaven.org aka vx.netlux.org</small></div>
<div style="margin-top: 2px; float: left;" class="adsapeu">
<script type="text/rocketscript">
<!--
var _acic={dataProvider:10};(function(){var e=document.createElement("script");e.type="text/javascript";e.async=true;e.src="//www.acint.net/aci.js";var t=document.getElementsByTagName("script")[0];t.parentNode.insertBefore(e,t)})()
//-->
</script>
</div>
<script data-rocketsrc="http://www.google-analytics.com/urchin.js" type="text/rocketscript"></script><script type="text/rocketscript">try { _uacct = "UA-590608-1"; urchinTracker(); } catch(err) {}</script>
<div style="display: none;"><a href="/lib/index.php?lang=de&amp;id=vsp37">de</a><a href="/lib/index.php?lang=en&amp;id=vsp37">en</a><a href="/lib/index.php?lang=es&amp;id=vsp37">es</a><a href="/lib/index.php?lang=it&amp;id=vsp37">it</a><a href="/lib/index.php?lang=fr&amp;id=vsp37">fr</a><a href="/lib/index.php?lang=pl&amp;id=vsp37">pl</a><a href="/lib/index.php?lang=ru&amp;id=vsp37">ru</a><a href="/lib/index.php?lang=ua&amp;id=vsp37">ua</a></div>
</body>
</html>
