<?xml version="1.0"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<title> SPTH 'Infection of biological DNA with digital Computer Code' (VX heaven)</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8"/>
<meta name="Author" content="SPTH"/>
<meta name="KeyWords" lang="en" content="computer virus, virus, virii,vx, компьютерные вирусы, вирус, вири, SPTH,Infection of biological DNA with digital Computer Code, fasta, code, bacteria, plain, life, digital, cell, watermarks, mycoplasma, computer, files, synthetic, artificial, venter, craig"/>
<meta name="Description" content="Biological life spreads in the biological-chemical world, computer codes can spread in the digital computer world. That is a rule - no self- replicator has ever overcome the digital-biological barrier. Until today. Here I show a method how a digital computer code can infect biological DNA, thus spread in the biological-chemical world. The method is mainly based on the fantastic research by the J. Craig Venter Institute on synthetic life, and might ask new questions about the definition of life itself."/>
<script type="text/javascript">
//<![CDATA[
try{if (!window.CloudFlare) {var CloudFlare=[{verbose:0,p:0,byc:0,owlid:"cf",bag2:1,mirage2:0,oracle:0,paths:{cloudflare:"/cdn-cgi/nexp/dok3v=1613a3a185/"},atok:"047a5bcbf67431883fc9ed25fba33612",petok:"d8880c5b3d90f7c8c63efabce8a2e15d7c369385-1498756506-1800",zone:"vxheaven.org",rocket:"a",apps:{}}];document.write('<script type="text/javascript" src="//ajax.cloudflare.com/cdn-cgi/nexp/dok3v=85b614c0f6/cloudflare.min.js"><'+'\/script>');}}catch(e){};
//]]>
</script>
<link rel="icon" href="/favicon.ico" type="image/x-icon"/>
<link rel="shortcut icon" href="/favicon.ico" type="image/x-icon"/>
<link rel="stylesheet" type="text/css" href="/style.css"/><link rel="canonical" href="http://vxheaven.org/lib/vsp48.html"/>
<script type="text/rocketscript" data-rocketsrc="https://apis.google.com/js/plusone.js">{"parsetags": "explicit"}</script>
</head>
<body bgcolor="#dbc8a0" text="#302000" link="#225599" vlink="#113366">
<div class="s1">
<div style="float:right;"><a href="/lib/index.php?tbs=1"><img src="/img/max.gif" alt="Maximize"/></a></div> <form id="lf" style="margin: 0; float: right;" method="get" action="/index.php"><input type="hidden" name="action" value="set"/><select name="lang" onchange="javascript:document.getElementById('lf').submit();"><option value="ru">Русский</option><option selected="selected" value="en">English</option><option value="ua">Українська</option><option value="de">Deutsch</option><option value="es">Español</option><option value="fr">Fran&ccedil;ais</option><option value="it">Italiano</option><option value="pl">Polski</option></select></form>
<div style="float: right;"><div id="plusone"></div></div>
<script type="text/rocketscript">gapi.plusone.render("plusone", {"size":"small","count":"true"});</script>
<div style="float: right;" class="addthis_toolbox addthis_default_style">
<script type="text/rocketscript">var addthis_config = { ui_click: true }</script>
<a style="text-decoration: none; font-size: 10pt;" href="/?action=addthis" class="addthis_button_compact">Bookmark</a>
<script type="text/rocketscript" data-rocketsrc="http://s7.addthis.com/js/250/addthis_widget.js#username=herm1t"></script>
</div>
<div style="float: right;">
<script type="text/rocketscript" data-rocketsrc="http://www.google.com/cse/brand?form=cse-search-box&amp;lang=en"></script>
<form action="/search.php" id="cse-search-box">
<input type="hidden" name="cx" value="002577580816726040001:z9_irkorydo"/>
<input type="hidden" name="cof" value="FORID:10"/>
<input type="hidden" name="ie" value="UTF-8"/>
<input type="text" name="q" size="12" value=" "/>
<input type="submit" name="sa" value="Search"/>
</form>
</div><h1><a href="/" style="text-decoration: none; color: #000000;">VX Heaven</a></h1>
<span class="nav"><a href="/lib/">Library</a> <a href="/vl.php">Collection</a> <a href="/src.php">Sources</a> <a href="/vx.php?id=eidx">Engines</a> <a href="/vx.php?id=tidx">Constructors</a> <a href="/vx.php?id=sidx">Simulators</a> <a href="/vx.php?id=uidx">Utilities</a> <a href="/links.php">Links</a> <a href="/donate.php" style="color: #706020" id="donate">Donate</a> <a href="/forum" style="text-decoration: underline;">Forum</a> </span><br clear="all"/>
</div>
<div class="s2"><h1>Infection of biological DNA with digital Computer Code</h1><p><a href="/lib/?lang=en&amp;author=SPTH"> SPTH</a><br/> <em><a href="/vx.php?fid=2016#f2016">Valhalla #4</a></em><br/> <em>November 2013</em></p><script type="text/rocketscript">var disqus_url = 'http://vxheaven.org/lib/vsp48.html';</script>[<a style="" href="/lib/?lang=EN&amp;index=AI#vsp48">Back to index</a>] [<a href="/lib/vsp48.html#disqus_thread">Comments</a>]<br/> 
<p>Biological life spreads in the biological-chemical world, computer codes can spread in the digital computer world. That is a rule - no self- replicator has ever overcome the digital-biological barrier. Until today. Here I show a method how a digital computer code can infect biological DNA, thus spread in the biological-chemical world. The method is mainly based on the fantastic research by the J. Craig Venter Institute on synthetic life, and might ask new questions about the definition of life itself.</p>
<ul>
<li><a href="#c1">1) Introduction</a></li>
<li><a href="#c2">2) Craig Venter's synthetical cell</a>
<ul>
<li><a href="#c21">2.1) General concept</a></li>
<li><a href="#c22">2.2) Watermarks and DNA encoding language</a></li>
</ul></li>
<li><a href="#c3">3) FASTA file format</a>
<ul>
<li><a href="#c31">3.1) plain fasta-files</a></li>
<li><a href="#c32">3.2) xml fasta-files</a></li>
</ul></li>
<li><a href="#c4">4) Infection scenario</a>
<ul>
<li><a href="#c41">4.1) Stealth technique in DNA</a></li>
</ul></li>
<li><a href="#c5">5) Conclusion</a></li>
<li><a href="#c6">References</a></li>
</ul>
<h2><a name="c1"></a>1) Introduction</h2>
<p>In 2010, the J. Craig Venter Institute (JCVI) reported the creation of a bacterial cell with a chemically synthesized genome [1]. They sequenced the DNA of a bacteria (M.mycoides), modified several parts of its DNA in the computer, synthetized the novel genome and transplanted it to a different bacteria's cell (M.capricolum). They observed the control of the cell only by the new DNA. For verification, they introduced artificial "watermarks" sequences (non-coded part of the DNA) to the genome, which contained among other things the names of the involved scientists (written in a specially designed DNA encoding alphabet). The artificially created genome was capable of continuous self-replication. They call their new artificial bacterial Mycoplasma mycoides JCVI-syn1.0.</p>
<p>This is in my opinion one of the greatest scientific achievement in recent years.</p>
<p>In this text I explain the implementation of a computer code that makes the step from the digital to the biological world. The computer code, written in C++, hosts the DNA sequence of M.mycoides JCVI-syn1.0. At runtime it acts as follows:</p>
<ul>
<li>Preparing the DNA sequence of M.mycoides JCVI-syn1.0 in the memory (with slightly modified watermarks).</li>
<li>Encoding own file-content in base32. The base32 code is then encoded in JCVI's DNA-encoded alphabet.</li>
<li>This representation of its digital form is then copied to a watermark of the bacteria's genome in memory. With this, a fully functional bacterial DNA sequence including the digital code is generated.</li>
<li>Next it searches for FASTA-files on the computer, which are text-based representations of DNA sequences, commonly used by many DNA sequence libraries.</li>
<li>For each FASTA-file, it replaces the original DNA with the bacterial DNA containing the digital form of the computer code.</li>
</ul>
<p>The code has a classical self-replication mechanism as well, to eventually end up on a computer in a microbiology-laboratory with the ability of creating DNA out of digital genomes (such as laboratories by the JCVI).</p>
<p>If the scientists are incautious, the computer code's genome (instead of the intented original DNA) might be written to the biological cell. The new cell will start replicating in the biological world, and with it the representation of the digital computer code.</p>
<h2><a name="c2"></a>2) Craig Venter's synthetical cell</h2>
<h3><a name="c21"></a>2.1) General concept</h3>
<p>The team of Craig Venter has demonstrated how to create bacteria controlled by artificially designed and synthesized DNA. For that, they used the sequenced DNA of a ~1 mega-base pair bacteria M.mycoides. They modified the genome on the computer - deactivated several genes, and introduced watermarks (artificial non-coding parts of the DNA). A company called Blue Heron sequenced 1000 bp fragments of the full DNA. With a three-step procedere, they assembled the full DNA. This was transplanted into an empty receiver cell of the bacteria M.capricolum.</p>
<p>Amazingly, the cell with the new genom booted up, and was able to self-replicate. To verify that the expected genome was replicating, they introduced special functionality to the watermarks which are visible with chemical methods.</p>
<p>In their article [1] they write: <em>"This work provides a proof of principle for producing cells based on computer-designed genome sequences. DNA sequencing of a cellular genome allows storage of the genetic instructions for life as a digital file."</em></p>
<p>The project describe here uses the method of their proof-of-principle.</p>
<h3><a name="c22"></a>2.2) Watermarks and DNA encoding language</h3>
<p>The watermarks are parts of the genome that are not translated into functional proteins. That means: They are part of the DNA, but have no functional effect on the behaviour of the cell.</p>
<p>The watermarks are represented by nucleotides A,C,G,T. JCVI developed an encoding technique from DNA to human letters. Three nucleotides (one codon) represent one letter or ascii symbol. With that encoding methode, they encode readable information into the cell: It contains the name of the involved scientists, philosophical quotes and one html-code with an e-mail adresse.</p>
<p>The encoding from codons to letters has never been documented explicitly, but can be deduced mainly from the implicit information given in the article. The known alphabet looks like this:</p>
<pre class="source">
    TAG = a        GCA = k        TCC = u        AGA = 4        CAC = /
    AGT = b        AAC = l        TTG = v        GCG = 5        CCA = =
    TTT = c        CAA = m        GTC = w        GCC = 6        CGA = .
    ATT = d        TGC = n        GGT = x        TAT = 7        GAG = !
    TAA = e        CGT = o        CAT = y        CGC = 8        CAG = :
    GGC = f        ACA = p        TGG = z        GTA = 9        GGA = "
    TAC = g        TTA = q        TCT = 0        ATA = space    GTG = ,
    TCA = h        CTA = r        CTT = 1        GGG = chr(10)  TCG = @
    CTG = i        GCT = s        ACT = 2        AGC = >        CCC = -
    GTT = j        TGA = t        AAT = 3        CGG = &lt;
</pre>
<p>Four watermarks have been introduced to the modified bacterial DNA in the computer. As an example, a part of the DNA sequence of one watermark is:</p>
<pre class="source">
    GCTTAATAAATATGATCACTGTGCTACGCTATATGCCGTTGAATATAGGCTATATGATC
    ATAACATATATAGCTATAAGTGATAAGTTCCTGAATATAGGCTATATGATCATAACATA
    TACAACTGTACTCATGAATAAGTTAACGA 
</pre>
<p>The sequence is divided into three-nucleotide parts (codons):</p>
<pre class="source">
    GCT TAA TAA ATA TGA TCA CTG TGC TAC GCT ATA TGC CGT TGA ATA 
    TAG GCT ATA TGA TCA TAA CAT ATA TAG CTA TAA GTG ATA AGT TCC
    TGA ATA TAG GCT ATA TGA TCA TAA CAT ATA CAA CTG TAC TCA TGA
    ATA AGT TAA CGA
</pre>
<p>We can see in the above list that GCT stands for "s", TAA stands for "e", ATA is a space, TGA stands for "t" ... and so on. In the end we can extract the sentence: <em>"see things not as they are, but as they might be."</em></p>
<p>Obviously we can also write in this encoding technique:</p>
<pre class="source">
"hello vxers!" -> TCA TAA AAC AAC CGT ATA TTG GGT TAA CTA GCT GAG  
</pre>
<p>The full structure of the alphabet is not known ,therefor only 49 out of 64 codon's representation are presented here. However all of them are used in the watermark (i.e. there is no biological reason for not using specific codons).</p>
<h2><a name="c3"></a>3) FASTA file format</h2>
<p>Fasta files are textbased representations of nucleotide sequences, commonly used in micro-biologic libraries. There are two fasta-file types that I will describe here. The first one is plain fasta-format (which usually have the file-extention .fasta or .fas. Both are available from the genome-database <a href="http://www.ncbi.nlm.nih.gov/">http://www.ncbi.nlm.nih.gov/</a>.</p>
<p>For example, if you want to see the DNA of <a href="http://www.ncbi.nlm.nih.gov/nuccore/296455217">Mycoplasma mycoides JCVI-syn1.0</a> or something more common: <a href="http://www.ncbi.nlm.nih.gov/nuccore/BA000007.2">E.coli</a>.</p>
<h3><a name="c31"></a>3.1) plain fasta-files</h3>
<p>The plain fasta-files have a small header, followed by a plain representation of the DNA in the nucleotide basis (A, T, G, C). Two examples:</p>
<h4>a) Mycoplasma mycoides JCVI-syn1.0</h4>
<p>This is about 1MB of data</p>
<pre class="source">     
>gi|296455217|gb|CP002027.1| Synthetic Mycoplasma mycoides JCVI-syn1.0 clone sMmYCp235-1, complete sequence
ATGAACGTAAACGATATTTTAAAAGAACTTAAACTAAGTTTAATGGCTAATAAAAATATTGATGAATCCG
TGTATAACGACTATATAAAGACAATAAATATTCATAAAAAGGGGTTTTCTGATTATATTGTTGTTGTTAA
ATCACAATTTGGTTTGTTAGCTATAAAACAGTTTCGTCAAACTATTGAAAATGAGATAAAAAATATTTTA
AAAGAACCTGTAAATATTAGTTTTACATACGAACAAGAATATAAAAAACAACTAGAAAAAGATGAATTAA
TTAATAAAGATCATTCTGATATCATTACTAAAAAAGTTAAAAAAACTAATGAAAACACTTTTGAAAATTT
...
</pre>
<center><small>Mycoplasma mycoides JCVI-syn1.0.fasta</small></center>
<h4>b) Escherichia coli</h4>
<p>This is about 5.5MB of data</p>
<pre class="source">       
>gi|47118301|dbj|BA000007.2| Escherichia coli O157:H7 str. Sakai DNA, complete genome
AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTCTCTGACAGCAGC
TTCTGAACTGGTTACCTGCCGTGAGTAAATTAAAATTTTATTGACTTAGGTCACTAAATACTTTAACCAA
TATAGGCATAGCGCACAGACAGATAAAAATTACAGAGTACACAACATCCATGAAACGCATTAGCACCACC
ATTACCACCACCATCACCACCACCATCACCATTACCATTACCACAGGTAACGGTGCGGGCTGACGCGTAC
AGGAAACACAGAAAAAAGCCCGCACCTGACAGTGCGGGCTTTTTTTTCGACCAAAGGTAACGAGGTAACA    
...
</pre>
<center><small>E.coli.fasta</small></center>
<h3><a name="c32"></a>3.2) xml fasta-files</h3>
<p>The second form is pure DNA aswell, however in a small xml-file. Two examples again:</p>
<div class="xml" style="font-family:monospace;color: #000066;  border: solid thin #c2c1b1; background: #d6d5c5;"><span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;?xml</span> <span style="color: #000066;">version</span>=<span style="color: #ff0000;">&quot;1.0&quot;</span><span style="color: #000000; font-weight: bold;">?&gt;</span></span><br/>
&nbsp;<span style="color: #00bbdd;">&lt;!DOCTYPE TSeqSet PUBLIC &quot;-//NCBI//NCBI TSeq/EN&quot; &quot;http://www.ncbi.nlm.nih.gov/dtd/NCBI_TSeq.dtd&quot;&gt;</span><br/>
&nbsp;<span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;TSeqSet<span style="color: #000000; font-weight: bold;">&gt;</span></span></span><br/>
<span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;TSeq<span style="color: #000000; font-weight: bold;">&gt;</span></span></span><br/>
&nbsp; <span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;TSeq_seqtype</span> <span style="color: #000066;">value</span>=<span style="color: #ff0000;">&quot;nucleotide&quot;</span><span style="color: #000000; font-weight: bold;">/&gt;</span></span><br/>
&nbsp; <span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;TSeq_gi<span style="color: #000000; font-weight: bold;">&gt;</span></span></span>296455217<span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;/TSeq_gi<span style="color: #000000; font-weight: bold;">&gt;</span></span></span><br/>
&nbsp; <span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;TSeq_accver<span style="color: #000000; font-weight: bold;">&gt;</span></span></span>CP002027.1<span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;/TSeq_accver<span style="color: #000000; font-weight: bold;">&gt;</span></span></span><br/>
&nbsp; <span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;TSeq_taxid<span style="color: #000000; font-weight: bold;">&gt;</span></span></span>766747<span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;/TSeq_taxid<span style="color: #000000; font-weight: bold;">&gt;</span></span></span><br/>
&nbsp; <span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;TSeq_orgname<span style="color: #000000; font-weight: bold;">&gt;</span></span></span>synthetic Mycoplasma mycoides JCVI-syn1.0<span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;/TSeq_orgname<span style="color: #000000; font-weight: bold;">&gt;</span></span></span><br/>
&nbsp; <span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;TSeq_defline<span style="color: #000000; font-weight: bold;">&gt;</span></span></span>Synthetic Mycoplasma mycoides JCVI-syn1.0 clone sMmYCp235-1, complete sequence<span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;/TSeq_defline<span style="color: #000000; font-weight: bold;">&gt;</span></span></span><br/>
&nbsp; <span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;TSeq_length<span style="color: #000000; font-weight: bold;">&gt;</span></span></span>1078809<span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;/TSeq_length<span style="color: #000000; font-weight: bold;">&gt;</span></span></span><br/>
&nbsp; <span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;TSeq_sequence<span style="color: #000000; font-weight: bold;">&gt;</span></span></span>ATGAACGTAAACGATATTTTAAAAGAACTTAAACTAAGTTTAATGGCTAATAAAAATATTGATGAATCCGTGTATAACGACTATATAAAGACAATAAATATTCATAAAAAGGGGTTTTCTGATTATATTGTTGTTGTTAAATCA...<span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;/TSeq_sequence<span style="color: #000000; font-weight: bold;">&gt;</span></span></span><br/>
<span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;/TSeq<span style="color: #000000; font-weight: bold;">&gt;</span></span></span><br/>
<br/>
<span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;/TSeqSet<span style="color: #000000; font-weight: bold;">&gt;</span></span></span><br/>
&nbsp;</div>
<center><small>Mycoplasma mycoides JCVI-syn1.0.fasta.xml</small></center>
<p>or E.coli again:</p>
<div class="xml" style="font-family:monospace;color: #000066;  border: solid thin #c2c1b1; background: #d6d5c5;"><span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;?xml</span> <span style="color: #000066;">version</span>=<span style="color: #ff0000;">&quot;1.0&quot;</span><span style="color: #000000; font-weight: bold;">?&gt;</span></span><br/>
&nbsp;<span style="color: #00bbdd;">&lt;!DOCTYPE TSeqSet PUBLIC &quot;-//NCBI//NCBI TSeq/EN&quot; &quot;http://www.ncbi.nlm.nih.gov/dtd/NCBI_TSeq.dtd&quot;&gt;</span><br/>
&nbsp;<span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;TSeqSet<span style="color: #000000; font-weight: bold;">&gt;</span></span></span><br/>
<span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;TSeq<span style="color: #000000; font-weight: bold;">&gt;</span></span></span><br/>
&nbsp; <span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;TSeq_seqtype</span> <span style="color: #000066;">value</span>=<span style="color: #ff0000;">&quot;nucleotide&quot;</span><span style="color: #000000; font-weight: bold;">/&gt;</span></span><br/>
&nbsp; <span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;TSeq_gi<span style="color: #000000; font-weight: bold;">&gt;</span></span></span>47118301<span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;/TSeq_gi<span style="color: #000000; font-weight: bold;">&gt;</span></span></span><br/>
&nbsp; <span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;TSeq_accver<span style="color: #000000; font-weight: bold;">&gt;</span></span></span>BA000007.2<span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;/TSeq_accver<span style="color: #000000; font-weight: bold;">&gt;</span></span></span><br/>
&nbsp; <span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;TSeq_taxid<span style="color: #000000; font-weight: bold;">&gt;</span></span></span>386585<span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;/TSeq_taxid<span style="color: #000000; font-weight: bold;">&gt;</span></span></span><br/>
&nbsp; <span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;TSeq_orgname<span style="color: #000000; font-weight: bold;">&gt;</span></span></span>Escherichia coli O157:H7 str. Sakai<span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;/TSeq_orgname<span style="color: #000000; font-weight: bold;">&gt;</span></span></span><br/>
&nbsp; <span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;TSeq_defline<span style="color: #000000; font-weight: bold;">&gt;</span></span></span>Escherichia coli O157:H7 str. Sakai DNA, complete genome<span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;/TSeq_defline<span style="color: #000000; font-weight: bold;">&gt;</span></span></span><br/>
&nbsp; <span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;TSeq_length<span style="color: #000000; font-weight: bold;">&gt;</span></span></span>5498450<span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;/TSeq_length<span style="color: #000000; font-weight: bold;">&gt;</span></span></span><br/>
&nbsp; <span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;TSeq_sequence<span style="color: #000000; font-weight: bold;">&gt;</span></span></span>AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTCTCTGACAGCAGCTTCTGAACTGGTTACCTGCCGTGAGTAAATTAAAATTTTATTGACTTAGGTCACTAAATACTTTAACCAATATA...<span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;/TSeq_sequence<span style="color: #000000; font-weight: bold;">&gt;</span></span></span><br/>
<span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;/TSeq<span style="color: #000000; font-weight: bold;">&gt;</span></span></span><br/>
<br/>
<span style="color: #009900;"><span style="color: #000000; font-weight: bold;">&lt;/TSeqSet<span style="color: #000000; font-weight: bold;">&gt;</span></span></span><br/>
&nbsp;</div>
<h2><a name="c4"></a>4) Infection scenario</h2>
<p>The strategy of this digitally and biologically self-replicating code is the following:</p>
<p>It starts as a digital computer file, and replicates itself via local networks, USB sticks and other removeable devices. There are two potential scenarios to step from the digital to the biological world:</p>
<ul>
<li>The self-replicating code might end up at a USB stick of a microbiologist. (S)he runs it unintentionally at a computer that host DNA sequences (stored in the common fasta-file format) which will be synthesized and transplanted to receiving cells (such as in the labs of JCVI). The computer code will find these fasta-files and replace their DNA sequences with the bacterial genome of M.mycoides. This genome contains a watermark with the DNA-representation of the file-content of the computer code. When the DNA files are synthesized, the computer code is synthesized aswell, and will continuously self-replicate in the biological world in the form of a bacteria.</li>
<li>In this scenario, the code gets to the computer of a genome library (such as NCBI, National Center for Biotechnology Information). The computer code will search for FASTA files and replace their DNA content with its own DNA code. The employee will unintentionally upload the computer code's DNA instead of the original DNA.
<p>Then - back in a laboratory like that from JCVI - scientists will download the modified DNA sequence. When they synthesize the wrong DNA sequence, the computer code lands in the bacterial cell again, again capable of continuously self-replicate in the biological world.</p></li>
</ul>
<p>There is a different interesting scenario: First, Mycoplasma mycoides bacteria are usually infecting cattles and goats. Imagine an unknown outbreak of the here presented bacteria. Goats or cattles would get sick, and microbiologists want to know the exact reason. They take samples of the infectious cells and sequence them in their laboratories. Now they see the DNA, and find out that the bacteria contains a rather big non-coded sequence - the watermark. They find this very unnatural and analyse the watermark, also by applying Craig Venter's DNA encoding alphabet (because it is very famous due to their first fascinating results). After decoding, they see that the code only contains a,b,c,d,e,f,g,h,i,j,k,l,m,n,o,p,q,r,s,t,u,v,w,x,y,z,2,3,4,5,6,7</p>
<p>This is a curious structure, they research a bit and see that its base32-encoding. They decode it, and see 'M','Z',0x90,0x0,... They immediatly see that its a windows executeable, and I guess would be surprised :)</p>
<h3><a name="c41"></a>4.1) Stealth technique in the DNA</h3>
<p>In their modified genome of M.mycoides JCVI-syn1.0, the JCVI-team introduces four watermarks. Every watermark contains a special sequence which is useed to test whether a cell has the intended genome or is a contamination (for example, from the receiving cell).</p>
<p>In the supplementary material of their article [1], they describe the exact representation of these sequences (primer). Each of the four watermarks contain one primer. When they perform a multiplex PCR, each watermark creats a specific characteristic.</p>
<p>In my code, I removed the total original content of all watermarks, except for the identified primer-sequence. As a result, when a team tests the bacteria cell with the representation of the digital code, it will have the same characteristic as their original designed DNA. Thus the computer code's DNA will pass this test.</p>
<h2><a name="c5"></a>5) Conclusion</h2>
<p>I've shown the implementation of a technique that allows a digital computer code to make the step to the biological world. This is done by infecting a DNA-file with the genome of a self-replicating biological bacteria. The bacteria's genome contains the digital code of the self-replicator in form of a base32-representation encoded via Craig Venter's DNA encoding alphabet. The biological bacteria will self-replicator in the biological world, and so will the representation of the digital computer code.</p>
<p>The outbreak-probability of such cross-domain infectors is very low. The researchers in [1] have made ethical studies, and I'm convinced that they came up with perfect protections against potential attacks as this.</p>
<p>Finally, digital self-replicators are usually not considered as a form of life, even they fulfill the most important characteristic of life: capability of self-replication and subject to evolution [2]. I wonder whether this computer code can count as a form of life - if so, I would call it <strong>Mycoplasma mycoides SPTH-syn1.0</strong> :)</p>
<div align="right">
Second Part To Hell<br/>
October 2013<br/>
http://spth.virii.lu/<br/>
<a class="__cf_email__" href="/cdn-cgi/l/email-protection" data-cfemail="e1929184938dcf95898e8c8092a1868c80888dcf828e8c">[email&#160;protected]</a><script data-cfhash='f9e31' type="text/rocketscript">/* <![CDATA[ */!function(t,e,r,n,c,a,p){try{t=document.currentScript||function(){for(t=document.getElementsByTagName('script'),e=t.length;e--;)if(t[e].getAttribute('data-cfhash'))return t[e]}();if(t&&(c=t.previousSibling)){p=t.parentNode;if(a=c.getAttribute('data-cfemail')){for(e='',r='0x'+a.substr(0,2)|0,n=2;a.length-n;n+=2)e+='%'+('0'+('0x'+a.substr(n,2)^r).toString(16)).slice(-2);p.replaceChild(document.createTextNode(decodeURIComponent(e)),c)}p.removeChild(t)}}catch(u){}}()/* ]]> */</script><br/>
twitter: @SPTHvx<br/>
</div>
<h2><a name="c6"></a>References</h2>
<ol>
<li>Daniel G. Gibson et al., "Creation of a Bacterial Cell Controlled by a Chemically Synthesized Genome", Science 329, 52 (2010).</li>
<li>SPTH, <a href="/lib/vsp26.html">"Taking the redpill: Artificial Evolution in native x86 systems"</a>, (2010); <a href="/lib/vsp37.html">"Imitation of Life: Advanced system for native Artificial Evolution"</a>, in valhalla#1 (2011).</li>
</ol>
<p>PS: Thanks to hh86 for motivation. Thanks to the JCVI-team for their awesome research, looking forward reading more discoveries on the boarder between dead and living material!</p>
<p>PPS: I'm not a microbiologist (or biologist at all). Even if I tried as hard as possible, I can not rule out that some assumptions might be wrong, some things I might have misunderstand. In any case, the main idea should be valid.</p>
[<a style="" href="/lib/?lang=EN&amp;index=AI#vsp48">Back to index</a>] [<a href="/lib/vsp48.html#disqus_thread">Comments</a>]<br/> <div id="disqus_thread"></div>
<script type="text/rocketscript">
        /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
        var disqus_shortname = 'vxheaven'; // required: replace example with your forum shortname

        /* * * DON'T EDIT BELOW THIS LINE * * */
        (function() {
            var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
            dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
            (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
        })();
    </script>
<noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript></div>
<div><small>By accessing, viewing, downloading or otherwise using this content you agree to be bound by the <a href="/agreement.php">Terms of Use</a>!</small> <small>vxheaven.org aka vx.netlux.org</small></div>
<div style="margin-top: 2px; float: left;" class="adsapeu">
<script type="text/rocketscript">
<!--
var _acic={dataProvider:10};(function(){var e=document.createElement("script");e.type="text/javascript";e.async=true;e.src="//www.acint.net/aci.js";var t=document.getElementsByTagName("script")[0];t.parentNode.insertBefore(e,t)})()
//-->
</script>
</div>
<script data-rocketsrc="http://www.google-analytics.com/urchin.js" type="text/rocketscript"></script><script type="text/rocketscript">try { _uacct = "UA-590608-1"; urchinTracker(); } catch(err) {}</script>
<div style="display: none;"><a href="/lib/index.php?lang=de&amp;id=vsp48">de</a><a href="/lib/index.php?lang=en&amp;id=vsp48">en</a><a href="/lib/index.php?lang=es&amp;id=vsp48">es</a><a href="/lib/index.php?lang=it&amp;id=vsp48">it</a><a href="/lib/index.php?lang=fr&amp;id=vsp48">fr</a><a href="/lib/index.php?lang=pl&amp;id=vsp48">pl</a><a href="/lib/index.php?lang=ru&amp;id=vsp48">ru</a><a href="/lib/index.php?lang=ua&amp;id=vsp48">ua</a></div>
</body>
</html>
